<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>
  Class: Gene
  
    &mdash; Documentation by YARD 0.9.34
  
</title>

  <link rel="stylesheet" href="css/style.css" type="text/css" />

  <link rel="stylesheet" href="css/common.css" type="text/css" />

<script type="text/javascript">
  pathId = "Gene";
  relpath = '';
</script>


  <script type="text/javascript" charset="utf-8" src="js/jquery.js"></script>

  <script type="text/javascript" charset="utf-8" src="js/app.js"></script>


  </head>
  <body>
    <div class="nav_wrap">
      <iframe id="nav" src="class_list.html?1"></iframe>
      <div id="resizer"></div>
    </div>

    <div id="main" tabindex="-1">
      <div id="header">
        <div id="menu">
  
    <a href="_index.html">Index (G)</a> &raquo;
    
    
    <span class="title">Gene</span>
  
</div>

        <div id="search">
  
    <a class="full_list_link" id="class_list_link"
        href="class_list.html">

        <svg width="24" height="24">
          <rect x="0" y="4" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="12" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="20" width="24" height="4" rx="1" ry="1"></rect>
        </svg>
    </a>
  
</div>
        <div class="clear"></div>
      </div>

      <div id="content"><h1>Class: Gene
  
  
  
</h1>
<div class="box_info">
  
  <dl>
    <dt>Inherits:</dt>
    <dd>
      <span class="inheritName">Object</span>
      
        <ul class="fullTree">
          <li>Object</li>
          
            <li class="next">Gene</li>
          
        </ul>
        <a href="#" class="inheritanceTree">show all</a>
      
    </dd>
  </dl>
  

  
  
  
  
  

  

  
  <dl>
    <dt>Defined in:</dt>
    <dd>gene.rb</dd>
  </dl>
  
</div>

<h2>Overview</h2><div class="docstring">
  <div class="discussion">
    
<h2 id="label-Summary">Summary</h2>

<p>The Gene class represents a gene and provides methods for retrieving gene information, extracting exons, checking exon sequences, and writing GFF (General Feature Format) files.</p>


  </div>
</div>
<div class="tags">
  

</div>



  <h2>Instance Attribute Summary <small><a href="#" class="summary_toggle">collapse</a></small></h2>
  <ul class="summary">
    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#embl-instance_method" title="#embl (instance method)">#<strong>embl</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Accessors for instance variables.</p>
</div></span>
  
</li>

    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#exons-instance_method" title="#exons (instance method)">#<strong>exons</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Accessors for instance variables.</p>
</div></span>
  
</li>

    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#gene_id-instance_method" title="#gene_id (instance method)">#<strong>gene_id</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Accessors for instance variables.</p>
</div></span>
  
</li>

    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#gff_array-instance_method" title="#gff_array (instance method)">#<strong>gff_array</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Accessors for instance variables.</p>
</div></span>
  
</li>

    
  </ul>




  
    <h2>
      Instance Method Summary
      <small><a href="#" class="summary_toggle">collapse</a></small>
    </h2>

    <ul class="summary">
      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#bring_exons-instance_method" title="#bring_exons (instance method)">#<strong>bring_exons</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Retrieves exons from the geneâ€™s EMBL information and creates ExonSequence objects.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#check_sequences-instance_method" title="#check_sequences (instance method)">#<strong>check_sequences</strong>  &#x21d2; Array&lt;String&gt; </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Checks sequences of exons for repeated patterns and updates the GFF array.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#create_ensembl-instance_method" title="#create_ensembl (instance method)">#<strong>create_ensembl</strong>  &#x21d2; Bio::EMBL </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Creates a Bio::EMBL object by fetching gene information from the Ensembl database.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#initialize-instance_method" title="#initialize (instance method)">#<strong>initialize</strong>(gene_id)  &#x21d2; Gene </a>
    

    
  </span>
  
  
    <span class="note title constructor">constructor</span>
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Initializes a new Gene instance with the given gene ID.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#write_gff-instance_method" title="#write_gff (instance method)">#<strong>write_gff</strong>(output_filename = &#39;output.gff3&#39;)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Writes the GFF array to a file in GFF3 format.</p>
</div></span>
  
</li>

      
    </ul>
  

<div id="constructor_details" class="method_details_list">
  <h2>Constructor Details</h2>
  
    <div class="method_details first">
  <h3 class="signature first" id="initialize-instance_method">
  
    #<strong>initialize</strong>(gene_id)  &#x21d2; <tt><span class='object_link'><a href="" title="Gene (class)">Gene</a></span></tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Initializes a new Gene instance with the given gene ID.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>gene_id</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The gene ID.</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


11
12
13
14
15
16</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'gene.rb', line 11</span>

<span class='kw'>def</span> <span class='id identifier rubyid_initialize'>initialize</span><span class='lparen'>(</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='rparen'>)</span>
  <span class='ivar'>@gene_id</span> <span class='op'>=</span> <span class='id identifier rubyid_gene_id'>gene_id</span><span class='period'>.</span><span class='id identifier rubyid_upcase'>upcase</span>
  <span class='ivar'>@embl</span> <span class='op'>=</span> <span class='id identifier rubyid_create_ensembl'>create_ensembl</span>
  <span class='ivar'>@exons</span> <span class='op'>=</span> <span class='lbrace'>{</span><span class='rbrace'>}</span>
  <span class='ivar'>@gff_array</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='rbracket'>]</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
  
</div>

  <div id="instance_attr_details" class="attr_details">
    <h2>Instance Attribute Details</h2>
    
      
      <span id="embl=-instance_method"></span>
      <div class="method_details first">
  <h3 class="signature first" id="embl-instance_method">
  
    #<strong>embl</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Accessors for instance variables</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


6
7
8</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'gene.rb', line 6</span>

<span class='kw'>def</span> <span class='id identifier rubyid_embl'>embl</span>
  <span class='ivar'>@embl</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      
      <span id="exons=-instance_method"></span>
      <div class="method_details ">
  <h3 class="signature " id="exons-instance_method">
  
    #<strong>exons</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Accessors for instance variables</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


6
7
8</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'gene.rb', line 6</span>

<span class='kw'>def</span> <span class='id identifier rubyid_exons'>exons</span>
  <span class='ivar'>@exons</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      
      <span id="gene_id=-instance_method"></span>
      <div class="method_details ">
  <h3 class="signature " id="gene_id-instance_method">
  
    #<strong>gene_id</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Accessors for instance variables</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


6
7
8</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'gene.rb', line 6</span>

<span class='kw'>def</span> <span class='id identifier rubyid_gene_id'>gene_id</span>
  <span class='ivar'>@gene_id</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      
      <span id="gff_array=-instance_method"></span>
      <div class="method_details ">
  <h3 class="signature " id="gff_array-instance_method">
  
    #<strong>gff_array</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Accessors for instance variables</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


6
7
8</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'gene.rb', line 6</span>

<span class='kw'>def</span> <span class='id identifier rubyid_gff_array'>gff_array</span>
  <span class='ivar'>@gff_array</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
  </div>


  <div id="instance_method_details" class="method_details_list">
    <h2>Instance Method Details</h2>

    
      <div class="method_details first">
  <h3 class="signature first" id="bring_exons-instance_method">
  
    #<strong>bring_exons</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Retrieves exons from the geneâ€™s EMBL information and creates ExonSequence objects.</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'gene.rb', line 40</span>

<span class='kw'>def</span> <span class='id identifier rubyid_bring_exons'>bring_exons</span>
  <span class='id identifier rubyid_filter_array'>filter_array</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='rbracket'>]</span>
  <span class='ivar'>@embl</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feat'>feat</span><span class='op'>|</span>
    <span class='kw'>next</span> <span class='kw'>unless</span> <span class='id identifier rubyid_feat'>feat</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>exon</span><span class='tstring_end'>&#39;</span></span>
    
    <span class='comment'># Check for random genes. Some genes are associated with other genes in EMBL. We don&#39;t know why
</span>    <span class='kw'>next</span> <span class='kw'>unless</span> <span class='id identifier rubyid_feat'>feat</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='period'>.</span><span class='id identifier rubyid_values'>values</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span><span class='period'>.</span><span class='id identifier rubyid_include?'>include?</span><span class='lparen'>(</span><span class='ivar'>@gene_id</span><span class='rparen'>)</span>

    <span class='comment'># Filtering for repeated exons
</span>    <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_filter_array'>filter_array</span><span class='period'>.</span><span class='id identifier rubyid_include?'>include?</span><span class='lparen'>(</span><span class='id identifier rubyid_feat'>feat</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='period'>.</span><span class='id identifier rubyid_values'>values</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span><span class='lbracket'>[</span><span class='op'>-</span><span class='int'>1</span><span class='rbracket'>]</span><span class='rparen'>)</span>
    <span class='id identifier rubyid_filter_array'>filter_array</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_feat'>feat</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='period'>.</span><span class='id identifier rubyid_values'>values</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span><span class='lbracket'>[</span><span class='op'>-</span><span class='int'>1</span><span class='rbracket'>]</span>

    <span class='comment'># Creating the @exons hash with the exons of each gene
</span>    <span class='id identifier rubyid_exon_id'>exon_id</span> <span class='op'>=</span> <span class='id identifier rubyid_feat'>feat</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='period'>.</span><span class='id identifier rubyid_values'>values</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span><span class='lbracket'>[</span><span class='int'>8</span><span class='op'>..</span><span class='rbracket'>]</span>
    <span class='id identifier rubyid_position'>position</span> <span class='op'>=</span> <span class='id identifier rubyid_feat'>feat</span><span class='period'>.</span><span class='id identifier rubyid_position'>position</span><span class='period'>.</span><span class='id identifier rubyid_scan'>scan</span><span class='lparen'>(</span><span class='tstring'><span class='regexp_beg'>/</span><span class='tstring_content'>\d+</span><span class='regexp_end'>/</span></span><span class='rparen'>)</span>
    <span class='id identifier rubyid_start_pos'>start_pos</span> <span class='op'>=</span> <span class='id identifier rubyid_position'>position</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span>
    <span class='id identifier rubyid_end_pos'>end_pos</span> <span class='op'>=</span> <span class='id identifier rubyid_position'>position</span><span class='lbracket'>[</span><span class='int'>1</span><span class='rbracket'>]</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span>
    <span class='id identifier rubyid_exon_seq'>exon_seq</span> <span class='op'>=</span> <span class='ivar'>@embl</span><span class='period'>.</span><span class='id identifier rubyid_naseq'>naseq</span><span class='lbracket'>[</span><span class='lparen'>(</span><span class='id identifier rubyid_start_pos'>start_pos</span> <span class='op'>-</span> <span class='int'>1</span><span class='rparen'>)</span><span class='op'>..</span><span class='id identifier rubyid_end_pos'>end_pos</span><span class='rbracket'>]</span>

    <span class='ivar'>@exons</span><span class='lbracket'>[</span><span class='id identifier rubyid_exon_id'>exon_id</span><span class='rbracket'>]</span> <span class='op'>=</span> <span class='const'><span class='object_link'><a href="ExonSequence.html" title="ExonSequence (class)">ExonSequence</a></span></span><span class='period'>.</span><span class='id identifier rubyid_new'><span class='object_link'><a href="ExonSequence.html#initialize-instance_method" title="ExonSequence#initialize (method)">new</a></span></span><span class='lparen'>(</span><span class='id identifier rubyid_exon_id'>exon_id</span><span class='comma'>,</span> <span class='id identifier rubyid_exon_seq'>exon_seq</span><span class='comma'>,</span> <span class='ivar'>@embl</span><span class='period'>.</span><span class='id identifier rubyid_accession'>accession</span><span class='comma'>,</span> <span class='id identifier rubyid_feat'>feat</span><span class='period'>.</span><span class='id identifier rubyid_position'>position</span><span class='rparen'>)</span>
  <span class='kw'>end</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="check_sequences-instance_method">
  
    #<strong>check_sequences</strong>  &#x21d2; <tt>Array&lt;String&gt;</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Checks sequences of exons for repeated patterns and updates the GFF array.</p>


  </div>
</div>
<div class="tags">
  
<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Array&lt;String&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The GFF array containing information about repeated patterns in exons.</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


66
67
68
69
70
71
72
73
74</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'gene.rb', line 66</span>

<span class='kw'>def</span> <span class='id identifier rubyid_check_sequences'>check_sequences</span>
  <span class='ivar'>@exons</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid__key'>_key</span><span class='comma'>,</span> <span class='id identifier rubyid_value'>value</span><span class='op'>|</span>
    <span class='id identifier rubyid_value'>value</span><span class='period'>.</span><span class='id identifier rubyid_check_ctt'>check_ctt</span>
    <span class='ivar'>@gff_array</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_value'>value</span><span class='period'>.</span><span class='id identifier rubyid_check_aag'>check_aag</span>
    <span class='ivar'>@gff_array</span><span class='period'>.</span><span class='id identifier rubyid_compact!'>compact!</span>
  <span class='kw'>end</span>

  <span class='kw'>return</span> <span class='ivar'>@gff_array</span> <span class='kw'>unless</span> <span class='ivar'>@gff_array</span><span class='period'>.</span><span class='id identifier rubyid_nil?'>nil?</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="create_ensembl-instance_method">
  
    #<strong>create_ensembl</strong>  &#x21d2; <tt>Bio::EMBL</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Creates a Bio::EMBL object by fetching gene information from the Ensembl database.</p>


  </div>
</div>
<div class="tags">
  
<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Bio::EMBL</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The Bio::EMBL object containing gene information.</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'gene.rb', line 21</span>

<span class='kw'>def</span> <span class='id identifier rubyid_create_ensembl'>create_ensembl</span>
  <span class='id identifier rubyid_address'>address</span> <span class='op'>=</span> <span class='const'>URI</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&amp;format=embl&amp;id=</span><span class='embexpr_beg'>#{</span><span class='ivar'>@gene_id</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='const'>Net</span><span class='op'>::</span><span class='const'>HTTP</span><span class='period'>.</span><span class='id identifier rubyid_get_response'>get_response</span><span class='lparen'>(</span><span class='id identifier rubyid_address'>address</span><span class='rparen'>)</span>

  <span class='comment'># Control check
</span>  <span class='kw'>unless</span> <span class='id identifier rubyid_response'>response</span><span class='period'>.</span><span class='id identifier rubyid_code'>code</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>200</span><span class='tstring_end'>&#39;</span></span>
    <span class='id identifier rubyid_abort'>abort</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>Gene </span><span class='embexpr_beg'>#{</span><span class='ivar'>@id</span><span class='embexpr_end'>}</span><span class='tstring_content'> not found in dbfetch database</span><span class='tstring_end'>&quot;</span></span> 
  <span class='kw'>end</span>

  <span class='comment'># Writing the embl file in a new directory
</span>  <span class='const'>FileUtils</span><span class='period'>.</span><span class='id identifier rubyid_mkdir_p'>mkdir_p</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>data</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span>
  <span class='const'>File</span><span class='period'>.</span><span class='id identifier rubyid_open'>open</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>data/</span><span class='embexpr_beg'>#{</span><span class='ivar'>@gene_id</span><span class='embexpr_end'>}</span><span class='tstring_content'>.embl</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>w</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_embl_file'>embl_file</span><span class='op'>|</span>
    <span class='id identifier rubyid_embl_file'>embl_file</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_response'>response</span><span class='period'>.</span><span class='id identifier rubyid_body'>body</span>
  <span class='kw'>end</span>

  <span class='kw'>return</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>EMBL</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='id identifier rubyid_response'>response</span><span class='period'>.</span><span class='id identifier rubyid_body'>body</span><span class='rparen'>)</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="write_gff-instance_method">
  
    #<strong>write_gff</strong>(output_filename = &#39;output.gff3&#39;)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Writes the GFF array to a file in GFF3 format.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>output_filename</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>&#39;output.gff3&#39;</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>The name of the output GFF file. Default is â€˜output.gff3â€™.</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


79
80
81
82
83</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'gene.rb', line 79</span>

<span class='kw'>def</span> <span class='id identifier rubyid_write_gff'>write_gff</span><span class='lparen'>(</span><span class='id identifier rubyid_output_filename'>output_filename</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>output.gff3</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span>
  <span class='const'>File</span><span class='period'>.</span><span class='id identifier rubyid_open'>open</span><span class='lparen'>(</span><span class='id identifier rubyid_output_filename'>output_filename</span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>a</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_file'>file</span><span class='op'>|</span>
    <span class='ivar'>@gff_array</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='lbrace'>{</span> <span class='op'>|</span><span class='id identifier rubyid_line'>line</span><span class='op'>|</span> <span class='id identifier rubyid_file'>file</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_line'>line</span> <span class='rbrace'>}</span>
  <span class='kw'>end</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
  </div>

</div>

      <div id="footer">
  Generated on Wed Dec  6 17:47:17 2023 by
  <a href="https://yardoc.org" title="Yay! A Ruby Documentation Tool" target="_parent">yard</a>
  0.9.34 (ruby-3.0.0).
</div>

    </div>
  </body>
</html>