{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":write_gff"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "require 'net/http'\n",
    "require 'bio'\n",
    "require 'fileutils'\n",
    "\n",
    "gene_ids = ['AT4g27030', 'AT5g54270', 'AT1g21400', 'AT2G46340']\n",
    "\n",
    "class Gene\n",
    "  attr_accessor :gene_id, :embl, :exons, :gff_array\n",
    "  \n",
    "  def initialize(gene_id)\n",
    "    @gene_id = gene_id.upcase\n",
    "    @embl = self.create_ensembl\n",
    "    @exons = {}\n",
    "    @gff_array = []\n",
    "  end\n",
    "\n",
    "  def create_ensembl\n",
    "    address = URI(\"http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&format=embl&id=#{@gene_id}\")\n",
    "    response = Net::HTTP.get_response(address)\n",
    "    # Control check\n",
    "    unless response.code == \"200\"\n",
    "        abort \"Gene #{@id} not found in dbfetch database\" # No se si lo de dbfetch esta bien\n",
    "    end\n",
    "    # Lo comento pq me distrae\n",
    "    # Show a portion of the record\n",
    "    #puts record[1..100]\n",
    "    \n",
    "    # Writing the embl file in a new directory\n",
    "    # Estaría bien, si tenemos timepo, ponerle alguna feature rollo que compruebe el directorio y pregunte o algo del palo\n",
    "    FileUtils.mkdir_p('data')\n",
    "    File.open(\"data/#{@gene_id}.embl\", \"w\") do |embl_file|\n",
    "        embl_file.puts response.body\n",
    "    end\n",
    "    # Lo quito de momento porque me distrae\n",
    "    # puts \"EMBL file created for #{@gene_id}: gene_embl/#{@gene_id}\"\n",
    "\n",
    "    # Devuelve ya el objeto embl\n",
    "    return Bio::EMBL.new(response.body)\n",
    "  end\n",
    "\n",
    "  def bring_exons\n",
    "    filter_array = []\n",
    "    @embl.features.each do |feat|\n",
    "      next unless feat.feature == 'exon'\n",
    "      \n",
    "      # Check for random genes. Some genes are associated to other genes in EMBL. We don know why\n",
    "      next unless  feat.assoc.values[0].include?(@gene_id)\n",
    "      # A mi me parece que lo lógico sería de aqui abajo guardar el exon_info y la sequence\n",
    "      # puts \"Feature #{feat.feature}\"\n",
    "      #puts \"Position #{feat.position}\"#.match(/(\\d+)\\.\\.(\\d+)/).to_s}\"\n",
    "      # puts \"Exon id: #{feat.assoc.values[0][8..]}\" \n",
    "      # puts \"Exon sequence: #{@embl.naseq.splice(feat.position)}\"\n",
    "\n",
    "      # Filtering for repeated exons\n",
    "      \n",
    "      next if filter_array.include?(feat.assoc.values[0][-1]) \n",
    "      \n",
    "      filter_array << feat.assoc.values[0][-1]\n",
    "\n",
    "      # Creating the @exons hash with the position and the sequence of each gene\n",
    "      position = feat.position\n",
    "      exon_id = feat.assoc.values[0][8..]\n",
    "      exon_seq =  @embl.naseq.splice(position)\n",
    "      \n",
    "      @exons[exon_id] = ExonSequence.new(exon_id, exon_seq,@embl.accession) \n",
    "      @exons[exon_id].get_position(position)\n",
    "    end\n",
    "  end\n",
    "  \n",
    "  def check_sequences\n",
    "    # Estaría bien que loprimero que hiciese esta fuese llamar a bring_exons\n",
    "    @exons.each do |key, value|\n",
    "      puts key\n",
    "      value.check_ctt\n",
    "      @gff_array << value.check_aag\n",
    "      @gff_array.compact! # Creo que si no hay aags puede dar nulos\n",
    "      \n",
    "      ## CHECKING FOR CTTCTT\n",
    "\n",
    "      # Esto lo comento de moento porque no se bie que hay que hacer\n",
    "      #@cttctt_features << get_ctt_seq(key, value)\n",
    "\n",
    "      ## CHECKING FOR AAGAAG\n",
    "      #@aagaag_features << get_aag_seq(key, value)\n",
    "      puts \n",
    "    end\n",
    "  end \n",
    "\n",
    "  def write_gff(output_filename='outputNICO.gff3')\n",
    "    File.open(output_filename, 'a') do |file|\n",
    "      # file.puts \"##gff-version 3\" # He tenido que quitar esto pq al estar puesto en append sale muchas veces\n",
    "      @gff_array.each do |line| file.puts line end\n",
    "    end\n",
    "  end\n",
    "    \n",
    "\n",
    "  # def get_ctt_seq(id, value)\n",
    "  #   ctt_match = value['sequence'].match(/cttctt/)\n",
    "  #   if ctt_match\n",
    "  #     puts \"cttctt sequence found in #{id}\"\n",
    "  #     start_coord = value['position'].split('..')[0].to_i + ctt_match.begin(0)\n",
    "  #     end_coord = value['position'].split('..')[0].to_i + ctt_match.end(0) - 1\n",
    "  #     new_feature = Bio::Feature.new('cttctt_repeat', \"#{start_coord}..#{end_coord}\")\n",
    "  #     unless new_feature.nil?\n",
    "  #       return create_feature(start_coord, end_coord)\n",
    "\n",
    "  #     end\n",
    "  #   else\n",
    "  #     puts \"cttctt sequence not found in #{id}\"\n",
    "  #   end\n",
    "  # end\n",
    "\n",
    "  # def create_feature(start_coord, end_coord)\n",
    "  # end\n",
    "\n",
    "  # def get_aag_seq(id, value)\n",
    "  #   aag_match = value['sequence'].match(/aagaag/)\n",
    "  #   if aag_match\n",
    "  #     puts \"aagaag sequence found in #{id}\"\n",
    "  #     start_coord = value['position'].split('..')[0].to_i + aag_match.begin(0)\n",
    "  #     end_coord = value['position'].split('..')[0].to_i + aag_match.end(0) - 1\n",
    "  #     new_feature = Bio::Feature.new('aagaag_repeat', \"#{start_coord}..#{end_coord}\")\n",
    "  #     unless new_feature.nil?\n",
    "  #       return new_feature\n",
    "  #     end\n",
    "  #   else\n",
    "  #     puts \"aagaag sequence not found in #{id}\"\n",
    "  #   end\n",
    "  # end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":check_aag"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class ExonSequence\n",
    "    attr_accessor :sequence, :exon_id, :localization, :ctt_reapeats_array, :gene_accesion\n",
    "\n",
    "    def initialize(id, seq, accesion)\n",
    "        @exon_id = id\n",
    "        @sequence = Bio::Sequence.auto(seq)\n",
    "        @gene_accesion = accesion\n",
    "        @sequence.features = []\n",
    "        @ctt_reapeats_array = []\n",
    "    end\n",
    "\n",
    "    def get_position(position)\n",
    "        complement_match = position.match(/complement\\((\\d+)\\.\\.(\\d+)\\)/)\n",
    "        if complement_match\n",
    "            starting_coord = complement_match[1]\n",
    "            ending_coord = complement_match[2]\n",
    "            polarity ='-'\n",
    "          puts starting_coord\n",
    "          puts ending_coord\n",
    "        else \n",
    "            starting_coord = position.split('..')[0]\n",
    "            ending_coord = position.split('..')[1]\n",
    "            polarity='+'\n",
    "        end\n",
    "\n",
    "        # He metido la localizacion como una Bio::Feature pero es que no les veo el sentidoa  como funcionan. Me parece super poco práctica\n",
    "        #feature = Bio::Feature.new('location', 'starting_position' => starting_coord, 'ending_position' => ending_coord, 'polarity' =>polarity )\n",
    "        #@sequence.features << feature\n",
    "        #@sequence.features.each do |ft| puts ft.position['ending_position'] end\n",
    "        \n",
    "        # Alternativa a la localizacion\n",
    "        @localization = {\n",
    "            'starting_position' => starting_coord.to_i,\n",
    "            'ending_position' => ending_coord.to_i,\n",
    "            'polarity' => polarity}\n",
    "    end\n",
    "\n",
    "    def check_ctt(sequence = @sequence)\n",
    "        #puts 'aagaagaagaag'.match(/(aag){2,}/)\n",
    "        #match = @sequence.match(/(ctt){2,}/)\n",
    "        #match = nil\n",
    "        #seq_prueba = 'aaaaacttcttaaaacttctt'\n",
    "        match = sequence.match(/(ctt){2,}/)\n",
    "        \n",
    "        if match\n",
    "            puts \"CTT repeat found in #{@exon_id}\"\n",
    "            repeat_stating_coord = @localization['starting_position']+ match.begin(0)\n",
    "            repeat_ending_coord = @localization['starting_position'] + match.end(0) - 1\n",
    "            ctt_id = \"ctt_repeat_#{@exon_id.split('.')[0]}_#{repeat_stating_coord}_#{repeat_ending_coord}\"\n",
    "            # Preguntar a jacob que es esto del unless\n",
    "            unless @ctt_reapeats_array.include?(ctt_id)\n",
    "                @ctt_reapeats_array << \"#{@gene_accesion}\\tcustom\\tctt_repeat\\t#{repeat_stating_coord}\\t#{repeat_ending_coord}\\t#{@localization['polarity']}\\t.\\tID=#{ctt_id}\"\n",
    "            end\n",
    "            sequence = sequence.sub('cttctt', 'aaaaaa')\n",
    "            \n",
    "        end \n",
    "    end\n",
    "\n",
    "    def check_aag(sequence = @sequence)\n",
    "        match = sequence.match(/(aag){2,}/)\n",
    "        \n",
    "        if match\n",
    "            puts \"CTT repeat found in #{@exon_id}\"\n",
    "            repeat_stating_coord = @localization['starting_position']+ match.begin(0)\n",
    "            repeat_ending_coord = @localization['starting_position'] + match.end(0) - 1\n",
    "            ctt_id = \"ctt_repeat_#{@exon_id.split('.')[0]}_#{repeat_stating_coord}_#{repeat_ending_coord}\"\n",
    "            # Preguntar a jacob que es esto del unless\n",
    "            unless @ctt_reapeats_array.include?(ctt_id)\n",
    "                @ctt_reapeats_array << \"#{@gene_accesion}\\tcustom\\tctt_repeat\\t#{repeat_stating_coord}\\t#{repeat_ending_coord}\\t#{@localization['polarity']}\\t.\\tID=#{ctt_id}\"\n",
    "                \n",
    "            end\n",
    "            sequence = sequence.sub('aagaag', 'aaaaaa')\n",
    "            self.check_aag(sequence)\n",
    "        else\n",
    "            unless @ctt_reapeats_array[0].nil? \n",
    "                return @ctt_reapeats_array\n",
    "            end\n",
    "        end\n",
    "      \n",
    "    end\n",
    "\n",
    "\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AT4G27030.1.exon1\n",
      "\n",
      "AT5G54270.1.exon1\n",
      "CTT repeat found in AT5G54270.1.exon1\n",
      "\n",
      "AT5G54270.1.exon3\n",
      "CTT repeat found in AT5G54270.1.exon3\n",
      "CTT repeat found in AT5G54270.1.exon3\n",
      "\n",
      "AT5G54270.1.exon2\n",
      "\n",
      "AT1G21400.3.exon1\n",
      "CTT repeat found in AT1G21400.3.exon1\n",
      "CTT repeat found in AT1G21400.3.exon1\n",
      "\n",
      "AT1G21400.2.exon7\n",
      "\n",
      "AT1G21400.6.exon6\n",
      "\n",
      "AT1G21400.4.exon2\n",
      "\n",
      "AT1G21400.2.exon4\n",
      "\n",
      "AT1G21400.2.exon3\n",
      "CTT repeat found in AT1G21400.2.exon3\n",
      "\n",
      "AT1G21400.2.exon5\n",
      "CTT repeat found in AT1G21400.2.exon5\n",
      "\n",
      "AT1G21400.6.exon8\n",
      "\n",
      "AT1G21400.1.exon9\n",
      "\n",
      "AT2G46340.2.exon2\n",
      "\n",
      "AT2G46340.2.exon3\n",
      "\n",
      "AT2G46340.1.exon8\n",
      "CTT repeat found in AT2G46340.1.exon8\n",
      "CTT repeat found in AT2G46340.1.exon8\n",
      "\n",
      "AT2G46340.2.exon1\n",
      "CTT repeat found in AT2G46340.2.exon1\n",
      "\n",
      "AT2G46340.2.exon7\n",
      "\n",
      "AT2G46340.2.exon4\n",
      "CTT repeat found in AT2G46340.2.exon4\n",
      "CTT repeat found in AT2G46340.2.exon4\n",
      "CTT repeat found in AT2G46340.2.exon4\n",
      "\n",
      "AT2G46340.2.exon6\n",
      "CTT repeat found in AT2G46340.2.exon6\n",
      "\n",
      "AT2G46340.2.exon5\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[\"AT4g27030\", \"AT5g54270\", \"AT1g21400\", \"AT2G46340\"]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_ids = [ 'AT2G46340']\n",
    "\n",
    "gene_ids.each do |gene|\n",
    "    prueba = Gene.new(gene)\n",
    "    prueba.bring_exons\n",
    "    prueba.check_sequences\n",
    "    prueba.write_gff('prueba2.gff3')\n",
    "end"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Ruby 3.0.0",
   "language": "ruby",
   "name": "ruby"
  },
  "language_info": {
   "file_extension": ".rb",
   "mimetype": "application/x-ruby",
   "name": "ruby",
   "version": "3.0.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
