{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "0563835b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":fetch_uniprot_id"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class Gene\n",
    "#Gene_ID\tGene_name\tmutant_phenotype\n",
    "  attr_accessor :geneid  \n",
    "  attr_accessor :name\n",
    "\n",
    "  @@all_genes = Hash.new\n",
    "  \n",
    "  def initialize(geneid:, name:, phenotype:)\n",
    "    @geneid = geneid\n",
    "    abort \"\\ngene id does not match AGI locus code.  Aborting\\n\" unless @geneid.match(/at\\dg\\d+/i) # case insensitive match -= put 'i' after regex\n",
    "    @name = name\n",
    "    @phenotype = phenotype\n",
    "    \n",
    "    @@all_genes[geneid] = self  # add to Class variable hash of all instances\n",
    "    @linked_genes = Hash.new # genes + chisq values for genes linked to this one (as gene objects - starts as empty)\n",
    "  end\n",
    "\n",
    "\n",
    "  def self.all_genes\n",
    "    return @@all_genes\n",
    "  end\n",
    "  \n",
    "  \n",
    "  def self.find_gene_by_id(id)  # lookiup function\n",
    "    @@all_genes.each do |geneid, gene|\n",
    "      return gene if geneid == id\n",
    "    end\n",
    "  end\n",
    " \n",
    "  \n",
    "####GET UNIPROT ID \n",
    " def uniprot_id\n",
    "    unless @uniprot_id\n",
    "      # Fetch UniProt ID using the UniProt REST API\n",
    "      response = fetch_uniprot_id(@geneid)\n",
    "      if response['Result'].any?\n",
    "        @uniprot_id = response['Result'][0]['id']\n",
    "      else\n",
    "        @uniprot_id = nil\n",
    "      end\n",
    "    end\n",
    "    @uniprot_id\n",
    "  end\n",
    "\n",
    "  private\n",
    "\n",
    "  def fetch_uniprot_id(agi_id)\n",
    "    base_url = 'https://www.ebi.ac.uk/proteins/api'\n",
    "    endpoint = \"/agi-locus/#{agi_id}\"\n",
    "    url = base_url + endpoint\n",
    "\n",
    "    uri = URI(url)\n",
    "    response = Net::HTTP.get(uri)\n",
    "    JSON.parse(response)\n",
    "  end\n",
    "\n",
    "  \n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "fa684b75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":fetch"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "require 'rest-client'  \n",
    "\n",
    "# Create a function called \"fetch\" that we can re-use everywhere in our code\n",
    "\n",
    "def fetch(url, headers = {accept: \"*/*\"}, user = \"\", pass=\"\")\n",
    "  response = RestClient::Request.execute({\n",
    "    method: :get,\n",
    "    url: url.to_s,\n",
    "    user: user,\n",
    "    password: pass,\n",
    "    headers: headers})\n",
    "  return response\n",
    "  \n",
    "  rescue RestClient::ExceptionWithResponse => e\n",
    "    $stderr.puts e.inspect\n",
    "    response = false\n",
    "    return response  # now we are returning 'False', and we will check that with an \\\"if\\\" statement in our main code\n",
    "  rescue RestClient::Exception => e\n",
    "    $stderr.puts e.inspect\n",
    "    response = false\n",
    "    return response  # now we are returning 'False', and we will check that with an \\\"if\\\" statement in our main code\n",
    "  rescue Exception => e\n",
    "    $stderr.puts e.inspect\n",
    "    response = false\n",
    "    return response  # now we are returning 'False', and we will check that with an \\\"if\\\" statement in our main code\n",
    "end \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "id": "835f6973",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":retrieve_genes_from_file"
      ]
     },
     "execution_count": 407,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "require 'rest-client'\n",
    "\n",
    "class Gene\n",
    "  @@genes = []\n",
    "  attr_accessor :id\n",
    "\n",
    "  def initialize(id:)\n",
    "    @id = id\n",
    "    @@genes << self\n",
    "  end\n",
    "\n",
    "  def self.all_genes\n",
    "    @@genes\n",
    "  end\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 416,
   "id": "b09f5321",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":all_genes"
      ]
     },
     "execution_count": 416,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "require 'rest-client'\n",
    "\n",
    "class AnnotatedGene\n",
    "  attr_reader :id, :protein_name\n",
    "\n",
    "  def initialize(id:, protein_name:)\n",
    "    @id = id\n",
    "    @protein_name = protein_name\n",
    "  end\n",
    "\n",
    "  def self.retrieve_genes_from_file(filename)\n",
    "    protein_ids = []\n",
    "\n",
    "    array_of_lines = IO.readlines(filename)\n",
    "    array_of_lines[1..].each do |data|\n",
    "      geneid = data.split(\"\\t\").first.chomp  # Remove newline character\n",
    "\n",
    "      res = RestClient.get(\"https://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&format=embl&id=#{geneid}&style=raw\")\n",
    "\n",
    "      unless res\n",
    "        abort \"failed to retrieve #{geneid}\"\n",
    "      end\n",
    "\n",
    "      record = res.body\n",
    "\n",
    "      if result = record.match(/db_xref=\"Uniprot\\/SWISSPROT\\:([^\"]+)\"/)\n",
    "        protein_id = result[1]\n",
    "        puts \"the protein ID of #{geneid} is #{protein_id}\"\n",
    "        protein_ids << protein_id\n",
    "      else\n",
    "        puts \"couldn't find the protein ID of #{geneid}\"\n",
    "      end\n",
    "    end\n",
    "\n",
    "    protein_ids\n",
    "  end\n",
    "\n",
    "  def self.all_genes\n",
    "    # Assuming you have a method that returns all annotated genes\n",
    "  end\n",
    "end\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 438,
   "id": "e56dfa20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AT4g27030\n",
      "AT5g54270\n",
      "AT1g21400\n",
      "AT5g19120\n",
      "AT2g13360\n",
      "AT4g05180\n",
      "AT1g22690\n",
      "AT2g45170\n",
      "AT4g09650\n",
      "AT5g55620\n",
      "AT1g31330\n",
      "AT1g80440\n",
      "AT3g28270\n",
      "AT4g12800\n",
      "AT5g04140\n",
      "AT5g64040\n",
      "AT1g29660\n",
      "AT1g15820\n",
      "AT1g64500\n",
      "AT1g03130\n",
      "AT2g20670\n",
      "AT4g17090\n",
      "AT1g23740\n",
      "AT2g46340\n",
      "AT4g17460\n",
      "AT1g07010\n",
      "AT1g32220\n",
      "AT2g03750\n",
      "AT3g47470\n",
      "AT4g12830\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[\"AT4g27030\\n\", \"AT5g54270\\n\", \"AT1g21400\\n\", \"AT5g19120\\n\", \"AT2g13360\\n\", \"AT4g05180\\n\", \"AT1g22690\\n\", \"AT2g45170\\n\", \"AT4g09650\\n\", \"AT5g55620\\n\", \"AT1g31330\\n\", \"AT1g80440\\n\", \"AT3g28270\\n\", \"AT4g12800\\n\", \"AT5g04140\\n\", \"AT5g64040\\n\", \"AT1g29660\\n\", \"AT1g15820\\n\", \"AT1g64500\\n\", \"AT1g03130\\n\", \"AT2g20670\\n\", \"AT4g17090\\n\", \"AT1g23740\\n\", \"AT2g46340\\n\", \"AT4g17460\\n\", \"AT1g07010\\n\", \"AT1g32220\\n\", \"AT2g03750\\n\", \"AT3g47470\\n\", \"AT4g12830\"]"
      ]
     },
     "execution_count": 438,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "require 'csv'\n",
    "\n",
    "# Specify the path to the TSV file\n",
    "\n",
    "filename = 'ArabidopsisSubNetwork_GeneList-head5.txt'\n",
    "array_of_lines = IO.readlines(filename)\n",
    "array_of_lines[1..].each do |data|\n",
    "geneid = data.split(\"\\t\").first.chomp\n",
    "puts geneid\n",
    "\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5706ba5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 428,
   "id": "4dd1c202",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the protein ID of AT4g27030 is Q9SZ42\n",
      "the protein ID of AT5g54270 is F4K0A6\n",
      "the protein ID of AT1g21400 is Q9LPL5\n",
      "couldn't find the protein ID of AT5g19120\n",
      "the protein ID of AT2g13360 is Q56YA5\n",
      "the protein ID of AT4g05180 is Q41932\n",
      "the protein ID of AT1g22690 is Q8GWK5\n",
      "the protein ID of AT2g45170 is Q8S926\n",
      "the protein ID of AT4g09650 is Q9SSS9\n",
      "couldn't find the protein ID of AT5g55620\n",
      "the protein ID of AT1g31330 is Q9SHE8\n",
      "the protein ID of AT1g80440 is Q9M8L2\n",
      "the protein ID of AT3g28270 is Q9LHD9\n",
      "the protein ID of AT4g12800 is Q9SUI4\n",
      "the protein ID of AT5g04140 is Q9ZNZ7\n",
      "the protein ID of AT5g64040 is Q8L7V3\n",
      "the protein ID of AT1g29660 is Q9C7N5\n",
      "couldn't find the protein ID of AT1g15820\n",
      "couldn't find the protein ID of AT1g64500\n",
      "the protein ID of AT1g03130 is Q9SA56\n",
      "couldn't find the protein ID of AT2g20670\n",
      "the protein ID of AT4g17090 is O23553\n",
      "the protein ID of AT1g23740 is Q9ZUC2\n",
      "the protein ID of AT2g46340 is Q9SYX2\n",
      "the protein ID of AT4g17460 is P46600\n",
      "the protein ID of AT1g07010 is Q8L774\n",
      "the protein ID of AT1g32220 is Q9FVR6\n",
      "the protein ID of AT2g03750 is Q8RV79\n",
      "the protein ID of AT3g47470 is Q9SN90\n",
      "couldn't find the protein ID of AT4g12830\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[\"Q9SZ42\", \"F4K0A6\", \"Q9LPL5\", \"Q56YA5\", \"Q41932\", \"Q8GWK5\", \"Q8S926\", \"Q9SSS9\", \"Q9SHE8\", \"Q9M8L2\", \"Q9LHD9\", \"Q9SUI4\", \"Q9ZNZ7\", \"Q8L7V3\", \"Q9C7N5\", \"Q9SA56\", \"O23553\", \"Q9ZUC2\", \"Q9SYX2\", \"P46600\", \"Q8L774\", \"Q9FVR6\", \"Q8RV79\", \"Q9SN90\"]"
      ]
     },
     "execution_count": 428,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Retrieve protein IDs from the file\n",
    "protein_ids = AnnotatedGene.retrieve_genes_from_file(filename)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 429,
   "id": "acd6f823",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"Q9SZ42\", \"F4K0A6\", \"Q9LPL5\", \"Q56YA5\", \"Q41932\", \"Q8GWK5\", \"Q8S926\", \"Q9SSS9\", \"Q9SHE8\", \"Q9M8L2\", \"Q9LHD9\", \"Q9SUI4\", \"Q9ZNZ7\", \"Q8L7V3\", \"Q9C7N5\", \"Q9SA56\", \"O23553\", \"Q9ZUC2\", \"Q9SYX2\", \"P46600\", \"Q8L774\", \"Q9FVR6\", \"Q8RV79\", \"Q9SN90\"]"
      ]
     },
     "execution_count": 429,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 420,
   "id": "34439d2b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "(irb):3: warning: already initialized constant #<Class:0x0000560b51d1f968>::InteractomeBuilder::PSICQUIC_BASE_URL\n",
      "(irb):3: warning: previous definition of PSICQUIC_BASE_URL was here\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       ":expand_interactome"
      ]
     },
     "execution_count": 420,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "require 'rest-client'\n",
    "\n",
    "class InteractomeBuilder\n",
    "  PSICQUIC_BASE_URL = 'http://www.ebi.ac.uk/Tools/webservices/psicquic/intact/webservices/current/search/query/'\n",
    "\n",
    "  def initialize(gene_list)\n",
    "    @gene_list = gene_list\n",
    "  end\n",
    "\n",
    "  def build_interactome\n",
    "    interactome = {}\n",
    "\n",
    "    @gene_list.each do |query_gene|\n",
    "      interactome[query_gene] = fetch_interactions(query_gene)\n",
    "    end\n",
    "\n",
    "    expand_interactome(interactome)\n",
    "  end\n",
    "\n",
    "  private\n",
    "\n",
    "  def fetch_interactions(query_gene, quality_threshold = 0.60)\n",
    "    url = \"#{PSICQUIC_BASE_URL}#{query_gene}?format=tab25\"\n",
    "    response = RestClient.get(url)\n",
    "\n",
    "    parse_interactions(response.body, query_gene, quality_threshold)\n",
    "  rescue RestClient::Exception => e\n",
    "    puts \"Error fetching interactions for #{query_gene}: #{e.message}\"\n",
    "    []\n",
    "  end\n",
    "\n",
    "   def parse_interactions(response_body, query_gene, quality_threshold)\n",
    "    interactions = []\n",
    "\n",
    "    response_body.lines.each do |line|\n",
    "      fields = line.strip.split(\"\\t\")\n",
    "      next if fields.empty?\n",
    "\n",
    "      intact_score_str = fields.last.match(/intact-miscore:(\\d+\\.\\d+)/)&.captures&.first\n",
    "      intact_score = intact_score_str.to_f if intact_score_str\n",
    "\n",
    "      next if intact_score.nil? || intact_score < quality_threshold\n",
    "\n",
    "      # Check if the query gene is in the first column\n",
    "      next unless fields[0] == \"uniprotkb:#{query_gene}\"\n",
    "\n",
    "      interacting_gene = fields[1].split(\":\").last\n",
    "\n",
    "      # Skip self-interactions\n",
    "      next if interacting_gene == query_gene\n",
    "\n",
    "      # Add interacting gene only if not already present\n",
    "      interactions << interacting_gene unless interactions.include?(interacting_gene)\n",
    "    end\n",
    "\n",
    "    interactions\n",
    "  end\n",
    "\n",
    "  def expand_interactome(interactome)\n",
    "    expanded_interactome = {}\n",
    "\n",
    "    interactome.each do |query_gene, interactions|\n",
    "      expanded_interactome[query_gene] = {}\n",
    "\n",
    "      interactions.each do |gene|\n",
    "        expanded_interactome[query_gene][gene] = fetch_interactions(gene).uniq\n",
    "      end\n",
    "    end\n",
    "\n",
    "    expanded_interactome\n",
    "  end\n",
    "end\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 537,
   "id": "7a9eb57e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**1st-level**: Q9SZ42 interacts with:\n",
      "**1st-level**: F4K0A6 interacts with:\n",
      "**1st-level**: Q9LPL5 interacts with:\n",
      "**1st-level**: Q56YA5 interacts with:\n",
      "**1st-level**: Q41932 interacts with:\n",
      "**1st-level**: Q8GWK5 interacts with:\n",
      "**1st-level**: Q8S926 interacts with:\n",
      "**1st-level**: Q9SSS9 interacts with:\n",
      "**1st-level**: Q9SHE8 interacts with:\n",
      "**1st-level**: Q9M8L2 interacts with:\n",
      "**1st-level**: Q9LHD9 interacts with:\n",
      "**1st-level**: Q9SUI4 interacts with:\n",
      "**1st-level**: Q9ZNZ7 interacts with:\n",
      "**1st-level**: Q8L7V3 interacts with:\n",
      "**1st-level**: Q9C7N5 interacts with:\n",
      "**1st-level**: Q9SA56 interacts with:\n",
      "**1st-level**: O23553 interacts with:\n",
      "**1st-level**: Q9ZUC2 interacts with:\n",
      "**1st-level**: Q9SYX2 interacts with:\n",
      "2nd-level:  P43254, which interacts with:  \n",
      "    3rd-level: Q9LJR3, Q94BM7, Q9SYX2, Q9SPL2, Q9FE22, Q39057, O24646, Q96524, Q43125\n",
      "2nd-level:  O50055, which interacts with:  \n",
      "    3rd-level: Q9SYX2, Q9LJR3, Q94BM7\n",
      "**1st-level**: P46600 interacts with:\n",
      "2nd-level:  O04292, which interacts with:  \n",
      "    3rd-level: P46600, Q9LEZ3, Q9SAD4\n",
      "**1st-level**: Q8L774 interacts with:\n",
      "**1st-level**: Q9FVR6 interacts with:\n",
      "**1st-level**: Q8RV79 interacts with:\n",
      "**1st-level**: Q9SN90 interacts with:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{\"Q9SZ42\"=>{}, \"F4K0A6\"=>{}, \"Q9LPL5\"=>{}, \"Q56YA5\"=>{}, \"Q41932\"=>{}, \"Q8GWK5\"=>{}, \"Q8S926\"=>{}, \"Q9SSS9\"=>{}, \"Q9SHE8\"=>{}, \"Q9M8L2\"=>{}, \"Q9LHD9\"=>{}, \"Q9SUI4\"=>{}, \"Q9ZNZ7\"=>{}, \"Q8L7V3\"=>{}, \"Q9C7N5\"=>{}, \"Q9SA56\"=>{}, \"O23553\"=>{}, \"Q9ZUC2\"=>{}, \"Q9SYX2\"=>{\"P43254\"=>[\"Q9LJR3\", \"Q94BM7\", \"Q9SYX2\", \"Q9SPL2\", \"Q9FE22\", \"Q39057\", \"O24646\", \"Q96524\", \"Q43125\"], \"O50055\"=>[\"Q9SYX2\", \"Q9LJR3\", \"Q94BM7\"]}, \"P46600\"=>{\"O04292\"=>[\"P46600\", \"Q9LEZ3\", \"Q9SAD4\"]}, \"Q8L774\"=>{}, \"Q9FVR6\"=>{}, \"Q8RV79\"=>{}, \"Q9SN90\"=>{}}"
      ]
     },
     "execution_count": 537,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example usage\n",
    "#gene_list = ['Q9SYX2', 'Q41932', 'Q94BM7','Q8GWK5','F4K0A6', 'Q41932','Q9LPL5','Q39057']\n",
    "interactome_builder = InteractomeBuilder.new(protein_ids)\n",
    "interactome = interactome_builder.build_interactome\n",
    "\n",
    "# Print the interactome (for demonstration purposes)\n",
    "interactome.each do |query_gene, second_level_interactions|\n",
    "  puts \"**1st-level**: #{query_gene} interacts with:\"\n",
    "  second_level_interactions.each do |second_gene, third_level_interactions|\n",
    "    puts \"2nd-level:  #{second_gene}, which interacts with:  \n",
    "    3rd-level: #{third_level_interactions.join(', ')}\"\n",
    "  end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 538,
   "id": "e27287bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of networks with more than one gene: 2\n"
     ]
    }
   ],
   "source": [
    "class NetworkCounter\n",
    "  def initialize(interactome)\n",
    "    @interactome = interactome\n",
    "    @visited_genes = []\n",
    "  end\n",
    "\n",
    "  def count_networks\n",
    "    networks = 0\n",
    "\n",
    "    @interactome.each do |query_gene, interactions|\n",
    "      next if interactions.empty?\n",
    "      next if @visited_genes.include?(query_gene) # Skip if the gene has already been visited\n",
    "\n",
    "      # Perform depth-first search to mark all genes in the same network\n",
    "      dfs(query_gene)\n",
    "      networks += 1\n",
    "    end\n",
    "\n",
    "    networks\n",
    "  end\n",
    "\n",
    "  private\n",
    "\n",
    "  def dfs(gene)\n",
    "    return if @visited_genes.include?(gene)\n",
    "\n",
    "    @visited_genes << gene\n",
    "\n",
    "    @interactome[gene]&.each do |neighbor_gene, _|\n",
    "      dfs(neighbor_gene)\n",
    "    end\n",
    "  end\n",
    "end\n",
    "\n",
    "# Example usage\n",
    "network_counter = NetworkCounter.new(interactome)\n",
    "networks = network_counter.count_networks\n",
    "\n",
    "puts \"Number of networks with more than one gene: #{networks}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 539,
   "id": "8f21f576",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":print_components"
      ]
     },
     "execution_count": 539,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "require 'rest-client'\n",
    "require 'json'\n",
    "\n",
    "class InteractomeProcessor\n",
    "  attr_reader :networks\n",
    "\n",
    "  def initialize(interactome)\n",
    "    @interactome = interactome\n",
    "    @lists = create_lists\n",
    "    @networks = []\n",
    "  end\n",
    "\n",
    "  def create_lists\n",
    "    lists = []\n",
    "\n",
    "    @interactome.each do |outer_key, inner_hash|\n",
    "      current_list = [outer_key]\n",
    "\n",
    "      inner_hash.each do |inner_key, inner_values|\n",
    "        current_list << inner_key\n",
    "        current_list.concat(inner_values)\n",
    "      end\n",
    "\n",
    "      lists << current_list\n",
    "    end\n",
    "\n",
    "    lists.reject! { |list| list.size <= 1 } # Remove lists with only one gene\n",
    "    lists\n",
    "  end\n",
    "\n",
    "  def join\n",
    "    joined_lists = []\n",
    "\n",
    "    @lists.each do |list1|\n",
    "      joined = false\n",
    "\n",
    "      joined_lists.each do |list2|\n",
    "        if (list1 & list2).any?\n",
    "          list2.concat(list1).uniq!\n",
    "          joined = true\n",
    "          break\n",
    "        end\n",
    "      end\n",
    "\n",
    "      joined_lists << list1 unless joined\n",
    "    end\n",
    "\n",
    "    joined_lists\n",
    "  end\n",
    "\n",
    "  def process\n",
    "    @networks = join\n",
    "    @networks.reject! { |network| network.size <= 1 } # Remove networks with only one gene\n",
    "  end\n",
    "\n",
    "  def print_network_report\n",
    "    puts \"#{@networks.size} networks have been identified:\"\n",
    "\n",
    "    @networks.each_with_index do |network, index|\n",
    "      puts \"Network #{index + 1}:\"\n",
    "      print_components(network)\n",
    "      print_go_terms(network)\n",
    "      puts \"KEGG terms:\"\n",
    "      # The KEGG annotations will be handled by InteractomeAnnotator\n",
    "      puts \"\\n\"\n",
    "    end\n",
    "  end\n",
    "\n",
    "  private\n",
    "\n",
    "  def print_components(network)\n",
    "    puts \"Components:\"\n",
    "    network.each do |gene|\n",
    "      puts \"#{gene}\"\n",
    "    end\n",
    "  end\n",
    "\n",
    "\n",
    "end\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 503,
   "id": "55861a25",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":print_kegg_terms"
      ]
     },
     "execution_count": 503,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "require 'rest-client'\n",
    "require 'json'\n",
    "\n",
    "class InteractomeAnnotator\n",
    "  def initialize(networks)\n",
    "    @networks = networks\n",
    "    @go_annotations = Hash.new { |hash, key| hash[key] = { count: 0, genes: [] } }\n",
    "    @kegg_annotations = Hash.new { |hash, key| hash[key] = { count: 0, genes: [] } }\n",
    "  end\n",
    "\n",
    "  def annotate_with_go_and_kegg\n",
    "    annotate_networks\n",
    "    print_network_report\n",
    "  end\n",
    "\n",
    "  private\n",
    "\n",
    "  def retrieve_go_annotations(protein_id)\n",
    "    address = \"http://togows.dbcls.jp/entry/uniprot/#{protein_id}/dr.json\"\n",
    "    response = RestClient::Request.execute(method: :get, url: address)\n",
    "    data = JSON.parse(response.body)\n",
    "\n",
    "    go_terms = data[0][\"GO\"] if data[0][\"GO\"]\n",
    "    go_terms.each_with_object({}) do |go, hash|\n",
    "      next unless (go[2] =~ /IDA:/) || (go[2] =~ /IMP:/)\n",
    "\n",
    "      go_id = go[0]\n",
    "      go_name = go[1]\n",
    "\n",
    "      hash[go_id] = go_name\n",
    "    end if go_terms.is_a?(Array)\n",
    "  rescue RestClient::Exception => e\n",
    "    puts \"Error retrieving GO annotations for #{protein_id}: #{e.message}\"\n",
    "    {}\n",
    "  end\n",
    "\n",
    "  def retrieve_kegg_annotations(gene_id)\n",
    "    address = \"http://togows.dbcls.jp/entry/uniprot/#{gene_id}/dr.json\"\n",
    "    response = RestClient::Request.execute(method: :get, url: address)\n",
    "    data = JSON.parse(response.body)\n",
    "    kegg_terms = data[0][\"KEGG\"] if data[0][\"KEGG\"]\n",
    "\n",
    "    kegg_terms.each_with_object([]) do |kegg_info, result|\n",
    "      kegg_id = kegg_info[0]\n",
    "\n",
    "      address = \"http://togows.org/entry/kegg-genes/#{kegg_id}/pathways.json\"\n",
    "\n",
    "      begin\n",
    "        response = RestClient::Request.execute(method: :get, url: address)\n",
    "        data = JSON.parse(response.body)\n",
    "\n",
    "        result << { 'id' => kegg_id, 'description' => data[0] } if data[0]&.any?\n",
    "      rescue RestClient::Exception => e\n",
    "        puts \"Error retrieving KEGG annotations for #{gene_id}: #{e.message}\"\n",
    "      end\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def annotate_gene(gene, network_index)\n",
    "    go_terms = retrieve_go_annotations(gene)\n",
    "\n",
    "    if go_terms.nil?\n",
    "      puts \"No GO annotations found for #{gene}.\"\n",
    "      return\n",
    "    end\n",
    "\n",
    "    go_terms.each do |go_id, go_name|\n",
    "      @go_annotations[go_id][:count] += 1\n",
    "      @go_annotations[go_id][:genes] << { gene: gene, network: network_index + 1, go_name: go_name }\n",
    "    end\n",
    "\n",
    "    kegg_terms = retrieve_kegg_annotations(gene)\n",
    "\n",
    "    if kegg_terms.nil?\n",
    "      puts \"No KEGG annotations found for #{gene}.\"\n",
    "      return\n",
    "    end\n",
    "\n",
    "    kegg_terms.each do |kegg_info|\n",
    "      kegg_id = kegg_info['id']\n",
    "      kegg_data = kegg_info['description']\n",
    "\n",
    "      @kegg_annotations[kegg_id][:count] += 1\n",
    "      @kegg_annotations[kegg_id][:genes] << { gene: gene, network: network_index + 1, kegg_data: kegg_data }\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def annotate_network(network, network_index)\n",
    "    network.each do |gene|\n",
    "      annotate_gene(gene, network_index)\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def annotate_networks\n",
    "    @networks.each_with_index do |network, index|\n",
    "      annotate_network(network, index)\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def print_network_report\n",
    "    puts \"Network Report:\"\n",
    "    puts \" \"\n",
    "    @networks.each_with_index do |network, index|\n",
    "      puts \"Network #{index + 1}:\"\n",
    "      print_components(network)\n",
    "      print_go_terms(network)\n",
    "      print_kegg_terms(network)\n",
    "      puts \"\\n\"\n",
    "    end\n",
    "  end\n",
    "\n",
    "  private\n",
    "\n",
    "  def print_components(network)\n",
    "    puts \"Components:\"\n",
    "    network.each do |gene|\n",
    "      puts \"#{gene}\"\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def print_go_terms(network)\n",
    "    puts \"GO terms:\"\n",
    "    sorted_go_annotations = @go_annotations.sort_by { |_, info| -info[:count] }\n",
    "\n",
    "    sorted_go_annotations.each do |go_id, info|\n",
    "      next unless info[:genes].any? { |gene_info| network.include?(gene_info[:gene]) }\n",
    "\n",
    "      puts \"#{go_id} (#{info[:count]} occurrences): #{info[:genes].first[:go_name]}\"\n",
    "      info[:genes].each do |gene_info|\n",
    "        next unless network.include?(gene_info[:gene])\n",
    "\n",
    "        puts \"  Gene: #{gene_info[:gene]}\"\n",
    "      end\n",
    "    end\n",
    "  end\n",
    "\n",
    "def print_kegg_terms(network)\n",
    "  puts \"KEGG terms:\"\n",
    "  @kegg_annotations.each do |kegg_id, info|\n",
    "    next unless info[:genes].any? { |gene_info| network.include?(gene_info[:gene]) }\n",
    "\n",
    "    puts \"#{kegg_id} (#{info[:count]} occurrences)\"\n",
    "    info[:genes].each do |gene_info|\n",
    "      next unless network.include?(gene_info[:gene])\n",
    "\n",
    "      puts \"  Protein: #{gene_info[:gene]}\"\n",
    "\n",
    "      kegg_data = gene_info[:kegg_data]\n",
    "      next if kegg_data.nil?\n",
    "\n",
    "      if kegg_data.is_a?(Array)\n",
    "        kegg_data.each do |pathway|\n",
    "          pathway_id = pathway['id']\n",
    "          pathway_description = pathway['description']\n",
    "          puts \"    Pathway: #{pathway_id} - #{pathway_description}\"\n",
    "        end\n",
    "      else\n",
    "        # Handle the case when kegg_data is a string\n",
    "        puts \"    Pathway: #{kegg_data}\"\n",
    "      end\n",
    "    end\n",
    "  end\n",
    "end\n",
    "\n",
    "\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 596,
   "id": "84f2c31c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":print_kegg_terms"
      ]
     },
     "execution_count": 596,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "require 'rest-client'\n",
    "require 'json'\n",
    "\n",
    "class InteractomeAnnotator\n",
    "  def initialize(networks)\n",
    "    @networks = networks\n",
    "    @go_annotations = Hash.new { |hash, key| hash[key] = { count: 0, genes: [] } }\n",
    "    @kegg_annotations = Hash.new { |hash, key| hash[key] = { count: 0, genes: [], pathway_id: nil, pathway_description: nil } }\n",
    "  end\n",
    "\n",
    "  def annotate_with_go_and_kegg\n",
    "    annotate_networks\n",
    "    print_network_report\n",
    "  end\n",
    "\n",
    "  private\n",
    "\n",
    "  def retrieve_go_annotations(protein_id)\n",
    "    address = \"http://togows.dbcls.jp/entry/uniprot/#{protein_id}/dr.json\"\n",
    "    response = RestClient::Request.execute(method: :get, url: address)\n",
    "    data = JSON.parse(response.body)\n",
    "\n",
    "    go_terms = data[0][\"GO\"] if data[0][\"GO\"]\n",
    "    go_terms.each_with_object({}) do |go, hash|\n",
    "      next unless (go[2] =~ /IDA:/) || (go[2] =~ /IMP:/)\n",
    "\n",
    "      go_id = go[0]\n",
    "      go_name = go[1]\n",
    "\n",
    "      hash[go_id] = go_name\n",
    "    end if go_terms.is_a?(Array)\n",
    "  rescue RestClient::Exception => e\n",
    "    puts \"Error retrieving GO annotations for #{protein_id}: #{e.message}\"\n",
    "    {}\n",
    "  end\n",
    "\n",
    "  def retrieve_kegg_annotations(gene_id)\n",
    "    puts \"Retrieving KEGG annotations for gene #{gene_id}\"\n",
    "\n",
    "    address = \"http://togows.dbcls.jp/entry/uniprot/#{gene_id}/dr.json\"\n",
    "    response = RestClient::Request.execute(method: :get, url: address)\n",
    "    data = JSON.parse(response.body)\n",
    "\n",
    "    if data[0].nil?\n",
    "      puts \"No KEGG annotations found for gene #{gene_id}.\"\n",
    "      return []\n",
    "    end\n",
    "\n",
    "    kegg_terms = data[0][\"KEGG\"] if data[0][\"KEGG\"]\n",
    "\n",
    "    kegg_annotations = kegg_terms.each_with_object([]) do |kegg_info, result|\n",
    "      kegg_id = kegg_info[0]\n",
    "      address = \"http://togows.org/entry/kegg-genes/#{kegg_id}/pathways.json\"\n",
    "\n",
    "      begin\n",
    "        response = RestClient::Request.execute({\n",
    "          method: :get,\n",
    "          url: address\n",
    "        })\n",
    "        data = JSON.parse(response.body)\n",
    "\n",
    "        if data[0]&.any?\n",
    "          data[0].each do |line|\n",
    "            pathway_id = line[0]\n",
    "            pathway_description = line[1]\n",
    "            result << { pathway_id: pathway_id, pathway_description: pathway_description }\n",
    "          end\n",
    "        end\n",
    "      rescue RestClient::Exception => e\n",
    "        puts \"Error retrieving KEGG annotations for #{gene_id}: #{e.message}\"\n",
    "      end\n",
    "    end\n",
    "\n",
    "    puts \"Retrieved KEGG annotations: #{kegg_annotations.inspect}\"\n",
    "    kegg_annotations\n",
    "  end\n",
    "\n",
    " def annotate_gene(gene, network_index)\n",
    "  go_terms = retrieve_go_annotations(gene)\n",
    "\n",
    "  if go_terms.nil?\n",
    "    puts \"No GO annotations found for #{gene}.\"\n",
    "    return\n",
    "  end\n",
    "\n",
    "  go_terms.each do |go_id, go_name|\n",
    "    @go_annotations[go_id][:count] += 1\n",
    "    @go_annotations[go_id][:genes] << { gene: gene, network: network_index + 1, go_name: go_name }\n",
    "  end\n",
    "\n",
    "  kegg_terms = retrieve_kegg_annotations(gene)\n",
    "\n",
    "  if kegg_terms.nil?\n",
    "    puts \"No KEGG annotations found for #{gene}.\"\n",
    "    return\n",
    "  end\n",
    "\n",
    "  kegg_terms.each do |kegg_entry|  # Change variable name to avoid conflict\n",
    "    kegg_id = kegg_entry[:id]\n",
    "    pathway_id = kegg_entry[:pathway_id]\n",
    "    pathway_description = kegg_entry[:pathway_description]\n",
    "\n",
    "    # Initialize @kegg_annotations for each KEGG ID\n",
    "    @kegg_annotations[kegg_id] ||= { count: 0, genes: [], pathway_id: nil, pathway_description: nil }\n",
    "\n",
    "    @kegg_annotations[kegg_id][:count] += 1\n",
    "    @kegg_annotations[kegg_id][:pathway_id] = pathway_id\n",
    "    @kegg_annotations[kegg_id][:pathway_description] = pathway_description\n",
    "    @kegg_annotations[kegg_id][:genes] << { gene: gene, network: network_index + 1 }\n",
    "  end\n",
    "end\n",
    "\n",
    "  def annotate_network(network, network_index)\n",
    "    network.each do |gene|\n",
    "      annotate_gene(gene, network_index)\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def annotate_networks\n",
    "    @networks.each_with_index do |network, index|\n",
    "      annotate_network(network, index)\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def print_network_report\n",
    "    puts \"Network Report:\"\n",
    "    puts \" \"\n",
    "    @networks.each_with_index do |network, index|\n",
    "      puts \"Network #{index + 1}:\"\n",
    "      print_components(network)\n",
    "      print_go_terms(network)\n",
    "      print_kegg_terms(network)\n",
    "      puts \"\\n\"\n",
    "    end\n",
    "  end\n",
    "\n",
    "  private\n",
    "\n",
    "  def print_components(network)\n",
    "    puts \"Components:\"\n",
    "    network.each do |gene|\n",
    "      puts \"#{gene}\"\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def print_go_terms(network)\n",
    "    puts \"GO terms:\"\n",
    "    sorted_go_annotations = @go_annotations.sort_by { |_, info| -info[:count] }\n",
    "\n",
    "    sorted_go_annotations.each do |go_id, info|\n",
    "      next unless info[:genes].any? { |gene_info| network.include?(gene_info[:gene]) }\n",
    "\n",
    "      puts \"#{go_id} (#{info[:count]} occurrences): #{info[:genes].first[:go_name]}\"\n",
    "      info[:genes].each do |gene_info|\n",
    "        next unless network.include?(gene_info[:gene])\n",
    "\n",
    "        puts \"  Gene: #{gene_info[:gene]}\"\n",
    "      end\n",
    "    end\n",
    "  end\n",
    "\n",
    "def print_kegg_terms(network)\n",
    "  puts \"KEGG terms:\"\n",
    "  sorted_kegg_annotations = @kegg_annotations.sort_by { |_, info| -info[:count] }\n",
    "\n",
    "  sorted_kegg_annotations.each do |pathway_id, info|\n",
    "    next unless info[:genes].any? { |gene_info| network.include?(gene_info[:gene]) }\n",
    "\n",
    "    if info[:pathway_id] && info[:pathway_description]\n",
    "      puts \"#{info[:pathway_id]} (#{info[:count]} occurrences): #{info[:pathway_description]}\"\n",
    "      info[:genes].each do |gene_info|\n",
    "        next unless network.include?(gene_info[:gene])\n",
    "\n",
    "        puts \"  Gene: #{gene_info[:gene]}\"\n",
    "      end\n",
    "    else\n",
    "      puts \"Invalid KEGG entry for pathway_id: #{info[:pathway_id]}\"\n",
    "    end\n",
    "  end\n",
    "end\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8aac8ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 networks have been identified:\n",
      "Network 1:\n",
      "Components:\n",
      "Q9SYX2\n",
      "P43254\n",
      "Q9LJR3\n",
      "Q94BM7\n",
      "Q9SYX2\n",
      "Q9SPL2\n",
      "Q9FE22\n",
      "Q39057\n",
      "O24646\n",
      "Q96524\n",
      "Q43125\n",
      "O50055\n",
      "Q9SYX2\n",
      "Q9LJR3\n",
      "Q94BM7\n",
      "GO terms:\n",
      "KEGG terms:\n",
      "\n",
      "Network 2:\n",
      "Components:\n",
      "P46600\n",
      "O04292\n",
      "P46600\n",
      "Q9LEZ3\n",
      "Q9SAD4\n",
      "GO terms:\n",
      "KEGG terms:\n",
      "\n",
      "Retrieving KEGG annotations for gene Q9SYX2\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene P43254\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04120\", :pathway_description=>\"Ubiquitin mediated proteolysis\"}, {:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q9LJR3\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q94BM7\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q9SYX2\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q9SPL2\n",
      "Retrieved KEGG annotations: []\n",
      "Retrieving KEGG annotations for gene Q9FE22\n",
      "Retrieved KEGG annotations: []\n",
      "Retrieving KEGG annotations for gene Q39057\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene O24646\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q96524\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q43125\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene O50055\n",
      "Retrieved KEGG annotations: []\n",
      "Retrieving KEGG annotations for gene Q9SYX2\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q9LJR3\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q94BM7\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene P46600\n",
      "Retrieved KEGG annotations: []\n",
      "Retrieving KEGG annotations for gene O04292\n"
     ]
    }
   ],
   "source": [
    "# Create an instance of InteractomeProcessor\n",
    "processor = InteractomeProcessor.new(interactome)\n",
    "\n",
    "# Process the interactome\n",
    "processor.process\n",
    "\n",
    "# Print the network report\n",
    "processor.print_network_report\n",
    "\n",
    "# Access the networks from the processor\n",
    "networks = processor.networks\n",
    "\n",
    "\n",
    "# Create an instance of InteractomeAnnotator\n",
    "annotator = InteractomeAnnotator.new(networks)\n",
    "\n",
    "# Annotate and print the network report\n",
    "annotator.annotate_with_go_and_kegg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 592,
   "id": "65a10249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":print_kegg_terms"
      ]
     },
     "execution_count": 592,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "require 'rest-client'\n",
    "require 'json'\n",
    "\n",
    "class InteractomeAnnotator\n",
    "  def initialize(networks)\n",
    "    @networks = networks\n",
    "    @go_annotations = Hash.new { |hash, key| hash[key] = { count: 0, genes: [] } }\n",
    "    @kegg_annotations = Hash.new { |hash, key| hash[key] = { count: 0, genes: [], pathway_id: nil, pathway_description: nil } }\n",
    "  end\n",
    "\n",
    "  def annotate_with_go_and_kegg\n",
    "    annotate_networks\n",
    "    print_network_report\n",
    "  end\n",
    "\n",
    "  private\n",
    "\n",
    "  def retrieve_go_annotations(protein_id)\n",
    "    address = \"http://togows.dbcls.jp/entry/uniprot/#{protein_id}/dr.json\"\n",
    "    response = RestClient::Request.execute(method: :get, url: address)\n",
    "    data = JSON.parse(response.body)\n",
    "\n",
    "    go_terms = data[0][\"GO\"] if data[0][\"GO\"]\n",
    "    go_terms.each_with_object({}) do |go, hash|\n",
    "      next unless (go[2] =~ /IDA:/) || (go[2] =~ /IMP:/)\n",
    "\n",
    "      go_id = go[0]\n",
    "      go_name = go[1]\n",
    "\n",
    "      hash[go_id] = go_name\n",
    "    end if go_terms.is_a?(Array)\n",
    "  rescue RestClient::Exception => e\n",
    "    puts \"Error retrieving GO annotations for #{protein_id}: #{e.message}\"\n",
    "    {}\n",
    "  end\n",
    "\n",
    "  def retrieve_kegg_annotations(gene_id)\n",
    "    puts \"Retrieving KEGG annotations for gene #{gene_id}\"\n",
    "\n",
    "    address = \"http://togows.dbcls.jp/entry/uniprot/#{gene_id}/dr.json\"\n",
    "    response = RestClient::Request.execute(method: :get, url: address)\n",
    "    data = JSON.parse(response.body)\n",
    "\n",
    "    if data[0].nil?\n",
    "      puts \"No KEGG annotations found for gene #{gene_id}.\"\n",
    "      return []\n",
    "    end\n",
    "\n",
    "    kegg_terms = data[0][\"KEGG\"] if data[0][\"KEGG\"]\n",
    "\n",
    "    kegg_annotations = kegg_terms.each_with_object([]) do |kegg_info, result|\n",
    "      kegg_id = kegg_info[0]\n",
    "      address = \"http://togows.org/entry/kegg-genes/#{kegg_id}/pathways.json\"\n",
    "\n",
    "      begin\n",
    "        response = RestClient::Request.execute({\n",
    "          method: :get,\n",
    "          url: address\n",
    "        })\n",
    "        data = JSON.parse(response.body)\n",
    "\n",
    "        if data[0]&.any?\n",
    "          data[0].each do |line|\n",
    "            pathway_id = line[0]\n",
    "            pathway_description = line[1]\n",
    "            result << { pathway_id: pathway_id, pathway_description: pathway_description }\n",
    "          end\n",
    "        end\n",
    "      rescue RestClient::Exception => e\n",
    "        puts \"Error retrieving KEGG annotations for #{gene_id}: #{e.message}\"\n",
    "      end\n",
    "    end\n",
    "\n",
    "    puts \"Retrieved KEGG annotations: #{kegg_annotations.inspect}\"\n",
    "    kegg_annotations\n",
    "  end\n",
    "\n",
    "  def annotate_gene(gene, network_index)\n",
    "    go_terms = retrieve_go_annotations(gene)\n",
    "\n",
    "    if go_terms.nil?\n",
    "      puts \"No GO annotations found for #{gene}.\"\n",
    "      return\n",
    "    end\n",
    "\n",
    "    go_terms.each do |go_id, go_name|\n",
    "      @go_annotations[go_id][:count] += 1\n",
    "      @go_annotations[go_id][:genes] << { gene: gene, network: network_index + 1, go_name: go_name }\n",
    "    end\n",
    "\n",
    "    kegg_terms = retrieve_kegg_annotations(gene)\n",
    "\n",
    "    if kegg_terms.nil?\n",
    "      puts \"No KEGG annotations found for #{gene}.\"\n",
    "      return\n",
    "    end\n",
    "\n",
    "    kegg_terms.each do |kegg_entry|\n",
    "      kegg_id = kegg_entry[:id]\n",
    "      pathway_id = kegg_entry[:pathway_id]\n",
    "      pathway_description = kegg_entry[:pathway_description]\n",
    "\n",
    "      @kegg_annotations[kegg_id] ||= { count: 0, genes: [], pathway_id: nil, pathway_description: nil }\n",
    "\n",
    "      @kegg_annotations[kegg_id][:count] += 1\n",
    "      @kegg_annotations[kegg_id][:pathway_id] = pathway_id\n",
    "      @kegg_annotations[kegg_id][:pathway_description] = pathway_description\n",
    "      @kegg_annotations[kegg_id][:genes] << { gene: gene, network: network_index + 1 }\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def annotate_network(network, network_index)\n",
    "    network.each do |gene|\n",
    "      annotate_gene(gene, network_index)\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def annotate_networks\n",
    "    @networks.each_with_index do |network, index|\n",
    "      annotate_network(network, index)\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def print_network_report\n",
    "    puts \"Network Report:\"\n",
    "    puts \" \"\n",
    "    @networks.each_with_index do |network, index|\n",
    "      puts \"Network #{index + 1}:\"\n",
    "      print_components(network)\n",
    "      print_go_terms(network)\n",
    "      print_kegg_terms(network)\n",
    "      puts \"\\n\"\n",
    "    end\n",
    "  end\n",
    "\n",
    "  private\n",
    "\n",
    "  def print_components(network)\n",
    "    puts \"Components:\"\n",
    "    network.each do |gene|\n",
    "      puts \"#{gene}\"\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def print_go_terms(network)\n",
    "    puts \"GO terms:\"\n",
    "    sorted_go_annotations = @go_annotations.sort_by { |_, info| -info[:count] }\n",
    "\n",
    "    sorted_go_annotations.each do |go_id, info|\n",
    "      next unless info[:genes].any? { |gene_info| network.include?(gene_info[:gene]) }\n",
    "\n",
    "      puts \"#{go_id} (#{info[:count]} occurrences): #{info[:genes].first[:go_name]}\"\n",
    "      info[:genes].each do |gene_info|\n",
    "        next unless network.include?(gene_info[:gene])\n",
    "\n",
    "        puts \"  Gene: #{gene_info[:gene]}\"\n",
    "      end\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def print_kegg_terms(network)\n",
    "    puts \"KEGG terms:\"\n",
    "    sorted_kegg_annotations = @kegg_annotations.sort_by { |_, info| -info[:count] }\n",
    "\n",
    "    sorted_kegg_annotations.each do |kegg_id, info|\n",
    "      next unless info[:genes].any? { |gene_info| network.include?(gene_info[:gene]) }\n",
    "\n",
    "      puts \"#{kegg_id} (#{info[:count]} occurrences): #{info[:genes].first[:pathway_description]}\"\n",
    "      info[:genes].each do |gene_info|\n",
    "        next unless network.include?(gene_info[:gene])\n",
    "\n",
    "        puts \"  Gene: #{gene_info[:gene]}\"\n",
    "      end\n",
    "    end\n",
    "  end\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 591,
   "id": "6bbad17e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving KEGG annotations for gene Q9SYX2\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene P43254\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04120\", :pathway_description=>\"Ubiquitin mediated proteolysis\"}, {:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q9LJR3\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q94BM7\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q9SPL2\n",
      "Retrieved KEGG annotations: []\n",
      "Retrieving KEGG annotations for gene Q9FE22\n",
      "Retrieved KEGG annotations: []\n",
      "Retrieving KEGG annotations for gene Q39057\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene O24646\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q96524\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q43125\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene O50055\n",
      "Retrieved KEGG annotations: []\n",
      "Retrieving KEGG annotations for gene P46600\n",
      "Retrieved KEGG annotations: []\n",
      "Retrieving KEGG annotations for gene O04292\n",
      "Retrieved KEGG annotations: []\n",
      "Retrieving KEGG annotations for gene P46600\n",
      "Retrieved KEGG annotations: []\n",
      "Retrieving KEGG annotations for gene Q9LEZ3\n",
      "Retrieved KEGG annotations: []\n",
      "Retrieving KEGG annotations for gene Q9SAD4\n",
      "Retrieved KEGG annotations: []\n",
      "Network Report:\n",
      " \n",
      "Network 1:\n",
      "Components:\n",
      "Q9SYX2\n",
      "P43254\n",
      "Q9LJR3\n",
      "Q94BM7\n",
      "Q9SPL2\n",
      "Q9FE22\n",
      "Q39057\n",
      "O24646\n",
      "Q96524\n",
      "Q43125\n",
      "O50055\n",
      "GO terms:\n",
      "GO:0005634 (10 occurrences): C:nucleus\n",
      "  Gene: Q9SYX2, Network: 1\n",
      "  Gene: P43254, Network: 1\n",
      "  Gene: Q9FE22, Network: 1\n",
      "  Gene: Q39057, Network: 1\n",
      "  Gene: O24646, Network: 1\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0016604 (4 occurrences): C:nuclear body\n",
      "  Gene: Q9SYX2, Network: 1\n",
      "  Gene: P43254, Network: 1\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0009637 (4 occurrences): P:response to blue light\n",
      "  Gene: Q9SYX2, Network: 1\n",
      "  Gene: Q94BM7, Network: 1\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0005737 (3 occurrences): C:cytoplasm\n",
      "  Gene: Q9SPL2, Network: 1\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0010218 (3 occurrences): P:response to far red light\n",
      "  Gene: Q9FE22, Network: 1\n",
      "  Gene: Q39057, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0009909 (3 occurrences): P:regulation of flower development\n",
      "  Gene: Q39057, Network: 1\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: O50055, Network: 1\n",
      "GO:0009640 (2 occurrences): P:photomorphogenesis\n",
      "  Gene: Q9SYX2, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0010017 (2 occurrences): P:red or far-red light signaling pathway\n",
      "  Gene: Q9SYX2, Network: 1\n",
      "  Gene: O24646, Network: 1\n",
      "GO:2000028 (2 occurrences): P:regulation of photoperiodism, flowering\n",
      "  Gene: Q9SYX2, Network: 1\n",
      "  Gene: Q96524, Network: 1\n",
      "GO:0046283 (2 occurrences): P:anthocyanin-containing compound metabolic process\n",
      "  Gene: P43254, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0009963 (2 occurrences): P:positive regulation of flavonoid biosynthetic process\n",
      "  Gene: P43254, Network: 1\n",
      "  Gene: O24646, Network: 1\n",
      "GO:0010224 (2 occurrences): P:response to UV-B\n",
      "  Gene: P43254, Network: 1\n",
      "  Gene: O24646, Network: 1\n",
      "GO:0005829 (2 occurrences): C:cytosol\n",
      "  Gene: Q9FE22, Network: 1\n",
      "GO:0005524 (2 occurrences): F:ATP binding\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0071949 (2 occurrences): F:FAD binding\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0010617 (2 occurrences): P:circadian regulation of calcium ion oscillation\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0072387 (2 occurrences): P:flavin adenine dinucleotide metabolic process\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0009638 (2 occurrences): P:phototropism\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0042752 (2 occurrences): P:regulation of circadian rhythm\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:1901371 (2 occurrences): P:regulation of leaf morphogenesis\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0009416 (2 occurrences): P:response to light stimulus\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0010244 (2 occurrences): P:response to low fluence blue light stimulus by blue low-fluence system\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:1902347 (2 occurrences): P:response to strigolactone\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0009658 (1 occurrences): P:chloroplast organization\n",
      "  Gene: Q9SYX2, Network: 1\n",
      "GO:0010100 (1 occurrences): P:negative regulation of photomorphogenesis\n",
      "  Gene: Q9SYX2, Network: 1\n",
      "GO:0048575 (1 occurrences): P:short-day photoperiodism, flowering\n",
      "  Gene: Q9SYX2, Network: 1\n",
      "GO:0061630 (1 occurrences): F:ubiquitin protein ligase activity\n",
      "  Gene: P43254, Network: 1\n",
      "GO:0004842 (1 occurrences): F:ubiquitin-protein transferase activity\n",
      "  Gene: P43254, Network: 1\n",
      "GO:0006281 (1 occurrences): P:DNA repair\n",
      "  Gene: P43254, Network: 1\n",
      "GO:0009649 (1 occurrences): P:entrainment of circadian clock\n",
      "  Gene: P43254, Network: 1\n",
      "GO:0048573 (1 occurrences): P:photoperiodism, flowering\n",
      "  Gene: P43254, Network: 1\n",
      "GO:0010119 (1 occurrences): P:regulation of stomatal movement\n",
      "  Gene: P43254, Network: 1\n",
      "GO:0009641 (1 occurrences): P:shade avoidance\n",
      "  Gene: P43254, Network: 1\n",
      "GO:0016567 (1 occurrences): P:protein ubiquitination\n",
      "  Gene: Q9SPL2, Network: 1\n",
      "GO:0003712 (1 occurrences): F:transcription coregulator activity\n",
      "  Gene: Q9FE22, Network: 1\n",
      "GO:0009738 (1 occurrences): P:abscisic acid-activated signaling pathway\n",
      "  Gene: Q9FE22, Network: 1\n",
      "GO:0003700 (1 occurrences): F:DNA-binding transcription factor activity\n",
      "  Gene: Q39057, Network: 1\n",
      "GO:0010018 (1 occurrences): P:far-red light signaling pathway\n",
      "  Gene: Q39057, Network: 1\n",
      "GO:0009908 (1 occurrences): P:flower development\n",
      "  Gene: Q39057, Network: 1\n",
      "GO:0032993 (1 occurrences): C:protein-DNA complex\n",
      "  Gene: O24646, Network: 1\n",
      "GO:0003677 (1 occurrences): F:DNA binding\n",
      "  Gene: O24646, Network: 1\n",
      "GO:0000976 (1 occurrences): F:transcription cis-regulatory region binding\n",
      "  Gene: O24646, Network: 1\n",
      "GO:0009740 (1 occurrences): P:gibberellic acid mediated signaling pathway\n",
      "  Gene: O24646, Network: 1\n",
      "GO:0009958 (1 occurrences): P:positive gravitropism\n",
      "  Gene: O24646, Network: 1\n",
      "GO:0031539 (1 occurrences): P:positive regulation of anthocyanin metabolic process\n",
      "  Gene: O24646, Network: 1\n",
      "GO:0042753 (1 occurrences): P:positive regulation of circadian rhythm\n",
      "  Gene: O24646, Network: 1\n",
      "GO:0009787 (1 occurrences): P:regulation of abscisic acid-activated signaling pathway\n",
      "  Gene: O24646, Network: 1\n",
      "GO:0010099 (1 occurrences): P:regulation of photomorphogenesis\n",
      "  Gene: O24646, Network: 1\n",
      "GO:0009737 (1 occurrences): P:response to abscisic acid\n",
      "  Gene: O24646, Network: 1\n",
      "GO:0006325 (1 occurrences): P:chromatin organization\n",
      "  Gene: Q96524, Network: 1\n",
      "GO:0006338 (1 occurrences): P:chromatin remodeling\n",
      "  Gene: Q96524, Network: 1\n",
      "GO:0051607 (1 occurrences): P:defense response to virus\n",
      "  Gene: Q96524, Network: 1\n",
      "GO:0048574 (1 occurrences): P:long-day photoperiodism, flowering\n",
      "  Gene: Q96524, Network: 1\n",
      "GO:0009911 (1 occurrences): P:positive regulation of flower development\n",
      "  Gene: Q96524, Network: 1\n",
      "GO:2000379 (1 occurrences): P:positive regulation of reactive oxygen species metabolic process\n",
      "  Gene: Q96524, Network: 1\n",
      "GO:0009882 (1 occurrences): F:blue light photoreceptor activity\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0016301 (1 occurrences): F:kinase activity\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0004672 (1 occurrences): F:protein kinase activity\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0060918 (1 occurrences): P:auxin transport\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0009583 (1 occurrences): P:detection of light stimulus\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:1901332 (1 occurrences): P:negative regulation of lateral root development\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Gene: Q43125, Network: 1\n",
      "GO:0010117 (1 occurrences): P:photoprotection\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0099402 (1 occurrences): P:plant organ development\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:1901529 (1 occurrences): P:positive regulation of anion channel activity\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:1900426 (1 occurrences): P:positive regulation of defense response to bacterium\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:1902448 (1 occurrences): P:positive regulation of shade avoidance\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:1901672 (1 occurrences): P:positive regulation of systemic acquired resistance\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0046777 (1 occurrences): P:protein autophosphorylation\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0010468 (1 occurrences): P:regulation of gene expression\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0010310 (1 occurrences): P:regulation of hydrogen peroxide metabolic process\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:2000377 (1 occurrences): P:regulation of reactive oxygen species metabolic process\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:2000652 (1 occurrences): P:regulation of secondary cell wall biogenesis\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0051510 (1 occurrences): P:regulation of unidimensional cell growth\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0009646 (1 occurrences): P:response to absence of light\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0009644 (1 occurrences): P:response to high light intensity\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0071000 (1 occurrences): P:response to magnetism\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0010114 (1 occurrences): P:response to red light\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0009266 (1 occurrences): P:response to temperature stimulus\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0010343 (1 occurrences): P:singlet oxygen-mediated programmed cell death\n",
      "  Gene: Q43125, Network: 1\n",
      "GO:0010118 (1 occurrences): P:stomatal movement\n",
      "  Gene: Q43125, Network: 1\n",
      "KEGG terms:\n",
      " (9 occurrences): \n",
      "  Gene: Q9SYX2, Network: 1\n",
      "  Gene: P43254, Network: 1\n",
      "  Gene: P43254, Network: 1\n",
      "  Gene: Q9LJR3, Network: 1\n",
      "  Gene: Q94BM7, Network: 1\n",
      "  Gene: Q39057, Network: 1\n",
      "  Gene: O24646, Network: 1\n",
      "  Gene: Q96524, Network: 1\n",
      "  Gene: Q43125, Network: 1\n",
      "\n",
      "Network 2:\n",
      "Components:\n",
      "P46600\n",
      "O04292\n",
      "P46600\n",
      "Q9LEZ3\n",
      "Q9SAD4\n",
      "GO terms:\n",
      "GO:0005634 (10 occurrences): C:nucleus\n",
      "  Gene: P46600, Network: 2\n",
      "  Gene: P46600, Network: 2\n",
      "  Gene: Q9SAD4, Network: 2\n",
      "GO:0005829 (2 occurrences): C:cytosol\n",
      "  Gene: Q9SAD4, Network: 2\n",
      "GO:0009943 (1 occurrences): P:adaxial/abaxial axis specification\n",
      "  Gene: O04292, Network: 2\n",
      "GO:0009855 (1 occurrences): P:determination of bilateral symmetry\n",
      "  Gene: O04292, Network: 2\n",
      "GO:0010014 (1 occurrences): P:meristem initiation\n",
      "  Gene: O04292, Network: 2\n",
      "GO:0009944 (1 occurrences): P:polarity specification of adaxial/abaxial axis\n",
      "  Gene: O04292, Network: 2\n",
      "GO:0010072 (1 occurrences): P:primary shoot apical meristem specification\n",
      "  Gene: O04292, Network: 2\n",
      "GO:0048825 (1 occurrences): P:cotyledon development\n",
      "  Gene: Q9SAD4, Network: 2\n",
      "GO:0009880 (1 occurrences): P:embryonic pattern specification\n",
      "  Gene: Q9SAD4, Network: 2\n",
      "GO:1905392 (1 occurrences): P:plant organ morphogenesis\n",
      "  Gene: Q9SAD4, Network: 2\n",
      "GO:0090708 (1 occurrences): P:specification of plant organ axis polarity\n",
      "  Gene: Q9SAD4, Network: 2\n",
      "KEGG terms:\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[[\"Q9SYX2\", \"P43254\", \"Q9LJR3\", \"Q94BM7\", \"Q9SPL2\", \"Q9FE22\", \"Q39057\", \"O24646\", \"Q96524\", \"Q43125\", \"O50055\"], [\"P46600\", \"O04292\", \"P46600\", \"Q9LEZ3\", \"Q9SAD4\"]]"
      ]
     },
     "execution_count": 591,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assuming you have an array of networks\n",
    "networks = [\n",
    "  [\"Q9SYX2\", \"P43254\", \"Q9LJR3\", \"Q94BM7\", \"Q9SPL2\", \"Q9FE22\", \"Q39057\", \"O24646\", \"Q96524\", \"Q43125\", \"O50055\"],\n",
    "  [\"P46600\", \"O04292\", \"P46600\", \"Q9LEZ3\", \"Q9SAD4\"]\n",
    "]\n",
    "\n",
    "# Create an instance of InteractomeAnnotator\n",
    "annotator = InteractomeAnnotator.new(networks)\n",
    "\n",
    "# Annotate and print the network report\n",
    "annotator.annotate_with_go_and_kegg\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 594,
   "id": "48c06475",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[\"Q9SYX2\", \"P43254\", \"Q9LJR3\", \"Q94BM7\", \"Q9SPL2\", \"Q9FE22\", \"Q39057\", \"O24646\", \"Q96524\", \"Q43125\", \"O50055\"]]"
      ]
     },
     "execution_count": 594,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interactome_processor.networks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 593,
   "id": "653fbc3d",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving KEGG annotations for gene Q9SYX2\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene P43254\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04120\", :pathway_description=>\"Ubiquitin mediated proteolysis\"}, {:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q9LJR3\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q94BM7\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q9SPL2\n",
      "Retrieved KEGG annotations: []\n",
      "Retrieving KEGG annotations for gene Q9FE22\n",
      "Retrieved KEGG annotations: []\n",
      "Retrieving KEGG annotations for gene Q39057\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene O24646\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q96524\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene Q43125\n",
      "Retrieved KEGG annotations: [{:pathway_id=>\"ath04712\", :pathway_description=>\"Circadian rhythm - plant\"}]\n",
      "Retrieving KEGG annotations for gene O50055\n",
      "Retrieved KEGG annotations: []\n",
      "Network Report:\n",
      " \n",
      "Network 1:\n",
      "Components:\n",
      "Q9SYX2\n",
      "P43254\n",
      "Q9LJR3\n",
      "Q94BM7\n",
      "Q9SPL2\n",
      "Q9FE22\n",
      "Q39057\n",
      "O24646\n",
      "Q96524\n",
      "Q43125\n",
      "O50055\n",
      "GO terms:\n",
      "GO:0005634 (7 occurrences): C:nucleus\n",
      "  Gene: Q9SYX2\n",
      "  Gene: P43254\n",
      "  Gene: Q9FE22\n",
      "  Gene: Q39057\n",
      "  Gene: O24646\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "GO:0016604 (4 occurrences): C:nuclear body\n",
      "  Gene: Q9SYX2\n",
      "  Gene: P43254\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "GO:0009637 (4 occurrences): P:response to blue light\n",
      "  Gene: Q9SYX2\n",
      "  Gene: Q94BM7\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "GO:0005737 (3 occurrences): C:cytoplasm\n",
      "  Gene: Q9SPL2\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "GO:0010218 (3 occurrences): P:response to far red light\n",
      "  Gene: Q9FE22\n",
      "  Gene: Q39057\n",
      "  Gene: Q43125\n",
      "GO:0009909 (3 occurrences): P:regulation of flower development\n",
      "  Gene: Q39057\n",
      "  Gene: Q96524\n",
      "  Gene: O50055\n",
      "GO:0009640 (2 occurrences): P:photomorphogenesis\n",
      "  Gene: Q9SYX2\n",
      "  Gene: Q43125\n",
      "GO:0010017 (2 occurrences): P:red or far-red light signaling pathway\n",
      "  Gene: Q9SYX2\n",
      "  Gene: O24646\n",
      "GO:2000028 (2 occurrences): P:regulation of photoperiodism, flowering\n",
      "  Gene: Q9SYX2\n",
      "  Gene: Q96524\n",
      "GO:0046283 (2 occurrences): P:anthocyanin-containing compound metabolic process\n",
      "  Gene: P43254\n",
      "  Gene: Q43125\n",
      "GO:0009963 (2 occurrences): P:positive regulation of flavonoid biosynthetic process\n",
      "  Gene: P43254\n",
      "  Gene: O24646\n",
      "GO:0010224 (2 occurrences): P:response to UV-B\n",
      "  Gene: P43254\n",
      "  Gene: O24646\n",
      "GO:0005524 (2 occurrences): F:ATP binding\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "GO:0071949 (2 occurrences): F:FAD binding\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "GO:0010617 (2 occurrences): P:circadian regulation of calcium ion oscillation\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "GO:0072387 (2 occurrences): P:flavin adenine dinucleotide metabolic process\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "GO:0009638 (2 occurrences): P:phototropism\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "GO:0042752 (2 occurrences): P:regulation of circadian rhythm\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "GO:1901371 (2 occurrences): P:regulation of leaf morphogenesis\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "GO:0009416 (2 occurrences): P:response to light stimulus\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "GO:0010244 (2 occurrences): P:response to low fluence blue light stimulus by blue low-fluence system\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "GO:1902347 (2 occurrences): P:response to strigolactone\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "GO:0009658 (1 occurrences): P:chloroplast organization\n",
      "  Gene: Q9SYX2\n",
      "GO:0010100 (1 occurrences): P:negative regulation of photomorphogenesis\n",
      "  Gene: Q9SYX2\n",
      "GO:0048575 (1 occurrences): P:short-day photoperiodism, flowering\n",
      "  Gene: Q9SYX2\n",
      "GO:0061630 (1 occurrences): F:ubiquitin protein ligase activity\n",
      "  Gene: P43254\n",
      "GO:0004842 (1 occurrences): F:ubiquitin-protein transferase activity\n",
      "  Gene: P43254\n",
      "GO:0006281 (1 occurrences): P:DNA repair\n",
      "  Gene: P43254\n",
      "GO:0009649 (1 occurrences): P:entrainment of circadian clock\n",
      "  Gene: P43254\n",
      "GO:0048573 (1 occurrences): P:photoperiodism, flowering\n",
      "  Gene: P43254\n",
      "GO:0010119 (1 occurrences): P:regulation of stomatal movement\n",
      "  Gene: P43254\n",
      "GO:0009641 (1 occurrences): P:shade avoidance\n",
      "  Gene: P43254\n",
      "GO:0016567 (1 occurrences): P:protein ubiquitination\n",
      "  Gene: Q9SPL2\n",
      "GO:0005829 (1 occurrences): C:cytosol\n",
      "  Gene: Q9FE22\n",
      "GO:0003712 (1 occurrences): F:transcription coregulator activity\n",
      "  Gene: Q9FE22\n",
      "GO:0009738 (1 occurrences): P:abscisic acid-activated signaling pathway\n",
      "  Gene: Q9FE22\n",
      "GO:0003700 (1 occurrences): F:DNA-binding transcription factor activity\n",
      "  Gene: Q39057\n",
      "GO:0010018 (1 occurrences): P:far-red light signaling pathway\n",
      "  Gene: Q39057\n",
      "GO:0009908 (1 occurrences): P:flower development\n",
      "  Gene: Q39057\n",
      "GO:0032993 (1 occurrences): C:protein-DNA complex\n",
      "  Gene: O24646\n",
      "GO:0003677 (1 occurrences): F:DNA binding\n",
      "  Gene: O24646\n",
      "GO:0000976 (1 occurrences): F:transcription cis-regulatory region binding\n",
      "  Gene: O24646\n",
      "GO:0009740 (1 occurrences): P:gibberellic acid mediated signaling pathway\n",
      "  Gene: O24646\n",
      "GO:0009958 (1 occurrences): P:positive gravitropism\n",
      "  Gene: O24646\n",
      "GO:0031539 (1 occurrences): P:positive regulation of anthocyanin metabolic process\n",
      "  Gene: O24646\n",
      "GO:0042753 (1 occurrences): P:positive regulation of circadian rhythm\n",
      "  Gene: O24646\n",
      "GO:0009787 (1 occurrences): P:regulation of abscisic acid-activated signaling pathway\n",
      "  Gene: O24646\n",
      "GO:0010099 (1 occurrences): P:regulation of photomorphogenesis\n",
      "  Gene: O24646\n",
      "GO:0009737 (1 occurrences): P:response to abscisic acid\n",
      "  Gene: O24646\n",
      "GO:0006325 (1 occurrences): P:chromatin organization\n",
      "  Gene: Q96524\n",
      "GO:0006338 (1 occurrences): P:chromatin remodeling\n",
      "  Gene: Q96524\n",
      "GO:0051607 (1 occurrences): P:defense response to virus\n",
      "  Gene: Q96524\n",
      "GO:0048574 (1 occurrences): P:long-day photoperiodism, flowering\n",
      "  Gene: Q96524\n",
      "GO:0009911 (1 occurrences): P:positive regulation of flower development\n",
      "  Gene: Q96524\n",
      "GO:2000379 (1 occurrences): P:positive regulation of reactive oxygen species metabolic process\n",
      "  Gene: Q96524\n",
      "GO:0009882 (1 occurrences): F:blue light photoreceptor activity\n",
      "  Gene: Q43125\n",
      "GO:0016301 (1 occurrences): F:kinase activity\n",
      "  Gene: Q43125\n",
      "GO:0004672 (1 occurrences): F:protein kinase activity\n",
      "  Gene: Q43125\n",
      "GO:0060918 (1 occurrences): P:auxin transport\n",
      "  Gene: Q43125\n",
      "GO:0009583 (1 occurrences): P:detection of light stimulus\n",
      "  Gene: Q43125\n",
      "GO:1901332 (1 occurrences): P:negative regulation of lateral root development\n",
      "  Gene: Q43125\n",
      "GO:0010117 (1 occurrences): P:photoprotection\n",
      "  Gene: Q43125\n",
      "GO:0099402 (1 occurrences): P:plant organ development\n",
      "  Gene: Q43125\n",
      "GO:1901529 (1 occurrences): P:positive regulation of anion channel activity\n",
      "  Gene: Q43125\n",
      "GO:1900426 (1 occurrences): P:positive regulation of defense response to bacterium\n",
      "  Gene: Q43125\n",
      "GO:1902448 (1 occurrences): P:positive regulation of shade avoidance\n",
      "  Gene: Q43125\n",
      "GO:1901672 (1 occurrences): P:positive regulation of systemic acquired resistance\n",
      "  Gene: Q43125\n",
      "GO:0046777 (1 occurrences): P:protein autophosphorylation\n",
      "  Gene: Q43125\n",
      "GO:0010468 (1 occurrences): P:regulation of gene expression\n",
      "  Gene: Q43125\n",
      "GO:0010310 (1 occurrences): P:regulation of hydrogen peroxide metabolic process\n",
      "  Gene: Q43125\n",
      "GO:2000377 (1 occurrences): P:regulation of reactive oxygen species metabolic process\n",
      "  Gene: Q43125\n",
      "GO:2000652 (1 occurrences): P:regulation of secondary cell wall biogenesis\n",
      "  Gene: Q43125\n",
      "GO:0051510 (1 occurrences): P:regulation of unidimensional cell growth\n",
      "  Gene: Q43125\n",
      "GO:0009646 (1 occurrences): P:response to absence of light\n",
      "  Gene: Q43125\n",
      "GO:0009644 (1 occurrences): P:response to high light intensity\n",
      "  Gene: Q43125\n",
      "GO:0071000 (1 occurrences): P:response to magnetism\n",
      "  Gene: Q43125\n",
      "GO:0010114 (1 occurrences): P:response to red light\n",
      "  Gene: Q43125\n",
      "GO:0009266 (1 occurrences): P:response to temperature stimulus\n",
      "  Gene: Q43125\n",
      "GO:0010343 (1 occurrences): P:singlet oxygen-mediated programmed cell death\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Gene: Q43125\n",
      "GO:0010118 (1 occurrences): P:stomatal movement\n",
      "  Gene: Q43125\n",
      "KEGG terms:\n",
      " (9 occurrences): \n",
      "  Gene: Q9SYX2\n",
      "  Gene: P43254\n",
      "  Gene: P43254\n",
      "  Gene: Q9LJR3\n",
      "  Gene: Q94BM7\n",
      "  Gene: Q39057\n",
      "  Gene: O24646\n",
      "  Gene: Q96524\n",
      "  Gene: Q43125\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[[\"Q9SYX2\", \"P43254\", \"Q9LJR3\", \"Q94BM7\", \"Q9SPL2\", \"Q9FE22\", \"Q39057\", \"O24646\", \"Q96524\", \"Q43125\", \"O50055\"]]"
      ]
     },
     "execution_count": 593,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotator = InteractomeAnnotator.new(processor.networks)\n",
    "annotator.annotate_with_go_and_kegg\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Ruby 3.0.0",
   "language": "ruby",
   "name": "ruby"
  },
  "language_info": {
   "file_extension": ".rb",
   "mimetype": "application/x-ruby",
   "name": "ruby",
   "version": "3.0.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
